---
title: "Examination of bundle sequences"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE)
```

```{r}
library(tidyverse)
library(ggtree)
library(treeio)
library(caper)
library(tidytree)
library(pegas)
library(phytools)
library(cowplot)
```

## Structure tree

```{r eval=TRUE, fig.height=6, fig.width=10}
hap_struc_table <- read_tsv("../combined_y1_y2_analyses/output/haplotype_all_structures.tsv") 
gene_count <- read_tsv("../combined_y1_y2_analyses/output/gene_locations_on_haplotypes.tsv") %>%
  count(contig, name, hap_struc_d) %>%
  pivot_wider(names_from = name, values_from = n, names_prefix = "AMY", values_fill = 0)  %>%
  mutate(amy_copy_number = str_c(AMY1, AMY2A, AMY2Ap, AMY2B, sep = "-")) %>%
  dplyr::select(-contig) %>%
  group_by_all() %>% 
  tally() %>%
  ungroup()
gene_count %>%
  group_by(hap_struc_d) %>%
  filter(n()>1)
structure_tree <- read.newick("../combined_y1_y2_analyses/output/haplotype_structures_tree.nwk")
structure_tree_grouped <- structure_tree %>%
  as_tibble() %>%
  left_join(gene_count, by=c("label"="hap_struc_d")) %>%
  groupClade(c(33, 37, 39, 42, 48, 51)) %>%
  mutate(group = as.character(group))
hap_group_table <- structure_tree_grouped %>% 
  dplyr::select(label, group) %>% 
  filter(!is.na(label))
group_color <- RColorBrewer::brewer.pal(6, "Set2")
amy_color <- MetBrewer::met.brewer("Klimt", 4, "discrete")
structure_ggtree <-  structure_tree_grouped %>%
  as.treedata() %>%
  ggtree(aes(color=group), linewidth=1) +
  scale_color_manual(values = c("black",group_color), guide="none") +
  geom_tiplab(align=TRUE, mapping = aes(label=label), size=3, offset = 0.03, color="black") +
  geom_tippoint(aes(size=n, fill=group), shape=21, color="black") +
  scale_fill_manual(values = group_color, guide="none") +
  geom_text2(aes(subset=!isTip, label=node), hjust=-.3, size=2) +
  xlim_tree(1.2)
structure_ggtree
```

```{r eval=TRUE, fig.height=6, fig.width=5, units = "in"}
structure_ggtree_tmp <-  structure_tree_grouped %>%
  as.treedata() %>%
  ggtree(aes(color=group), linewidth=1) +
  scale_color_manual(values = c("black",group_color), guide="none") +
  geom_tippoint(aes(size=n, fill=group), shape=21, color="black") +
  scale_fill_manual(values = group_color, guide="none") +
  #geom_text2(aes(subset=!isTip, label=node), hjust=-.3, size=2) +
  xlim_tree(0.33)
structure_ggtree <- structure_ggtree_tmp +
  geom_tiplab(align=TRUE, mapping = aes(label=amy_copy_number), size=3, offset = 0.04, color="black")
#nice example where this came from: https://www.janknappe.com/blog/r-irish-elections-gender/
histo_dots <- gene_count %>%
  dplyr::select(1:5) %>%
  pivot_longer(2:5, values_to = "n", names_to = "name") %>%
  uncount(n) %>%
  arrange(hap_struc_d, name) %>%
  group_by(hap_struc_d) %>%
  mutate(index=row_number()) %>%
  ungroup()
histo_dots_plot <- histo_dots %>%
  ggplot(aes(x=index, y=hap_struc_d, fill=name, shape=name)) +
  geom_point(size=3) +
  scale_shape_manual(values = 22:25) +
  scale_fill_manual(values = amy_color) +
  ggnewscale::new_scale_fill() +
  theme_void() +
  theme(axis.text.y = element_blank())
structure_tree_with_gene_count <- histo_dots_plot %>% aplot::insert_left(structure_ggtree, width=1.1)
structure_tree_with_gene_count
structure_ggtree_legend <- structure_ggtree_tmp +
  theme(legend.position=c(0.1, 0.93))
```

```{r eval=FALSE}
ggsave("figures/structure_tree_with_gene_count.pdf", structure_tree_with_gene_count, width = 5.5, height = 6, units = "in")
ggsave("figures/structure_tree_legend.pdf", structure_ggtree_legend, width = 2, height = 6, units = "in")
```

## Extract all bundle sequences from the fasta file

```{r eval=TRUE, fig.width=10, fig.height=5}
bundle_loc <- read_tsv("../combined_y1_y2_analyses/output/pgrtk/AMY_48_56_4_1000.bed", col_names = c("chrom", "start", "end", "tmp"), skip=1) %>%
  separate(tmp, into = c("name", "tmp1", "strand", "tmp2"), extra = "merge") %>%
  dplyr::select(-tmp1, -tmp2) %>%
  mutate(length=end-start) %>%
  group_by(name) %>%
  mutate(mean_length=mean(length), sd_length=sd(length)) %>%
  ungroup()

bundle_loc %>%
  ggplot()+
  geom_histogram(aes(x=length/1000, fill=length <= mean_length / 2)) +
  facet_wrap(~name, scales = "free", nrow = 2) +
  cowplot::theme_cowplot() +
  theme(legend.position = 'none')

bundle_loc_filtered <-  bundle_loc %>%
  filter(length > mean_length / 2) 

bundle_loc_filtered %>%
  count(chrom, name) %>%
  count(name, n) %>%
  pivot_wider(names_from=n, values_from = nn) %>%
  rename(bundle_name = name)

bundle_loc_filtered %>%
  group_by(name) %>%
  summarise(n=n(), mean_length=mean(length)) %>%
  rename(bundle_name = name)
```

```{r eval=FALSE}
bundle_loc_filtered %>%
  dplyr::select(chrom, start, end, name, strand) %>%
  write_tsv("bundle_locations_on_haplotypes_filtered.tsv")
```

These confirm that `tname` =`contig` = `chrom`, `tstart`=`start`, `tend=end`, and the combination of `chrome`, `start`, and `end` are unique for each gene.

There are a total of 89 chromosomes, matching the fasta file. I had to subset the sequence to 10000bp if they are longer because MUSCLE cannot handle longer sequences. I'll need to explore other tools or mapping to a reference

```{bash eval=FALSE}
cd /global/scratch/users/nicolas931010/amylase_diversity_project/HPRC_AMY_Sequences/bundle_tree
rm fasta/*
rm kalign/*
rm iqtree/*
#gunzip -c ../combined_y1_y2_analyses/input/combined_input/AMY1A_region_seq.fa.gz > AMY1A_region_seq.fa
SEQ=AMY1A_region_seq.fa
BUNDLE_LOC=bundle_locations_on_haplotypes_filtered.tsv
NLINE=`wc -l $BUNDLE_LOC | cut -f 1 -d ' '`
for I in `seq 2 $NLINE`; do
CHRM=`awk -v i=$I '{if(NR==i) print $1}' $BUNDLE_LOC`
START=`awk -v i=$I '{if(NR==i) print $2}' $BUNDLE_LOC`
END=`awk -v i=$I '{if(NR==i) print $3}' $BUNDLE_LOC`
BUNDLE=`awk -v i=$I '{if(NR==i) print $4}' $BUNDLE_LOC`
STRAND=`awk -v i=$I '{if(NR==i) print $5}' $BUNDLE_LOC`
OUT=fasta/bundle${BUNDLE}.fa
NAME=${CHRM}.${START}_${END}
echo '>'${NAME} >> ${OUT}
if [ $STRAND = 0 ]; then
echo 0
grep -A 1 ${CHRM} ${SEQ} | tail -n 1 | cut -c ${START}-${END} >> ${OUT}
else
echo 1
grep -A 1 ${CHRM} ${SEQ} | tail -n 1 | cut -c ${START}-${END} | tr ACGTacgt TGCAtgca | rev >> ${OUT}
fi
done
```

## Extract Neanderthal and Denisovan sequences as outgroups

```{bash eval=FALSE}
mkdir /global/scratch/users/nicolas931010/amylase_diversity_project/HPRC_AMY_Sequences/bundle_tree/outgroup
cd /global/scratch/users/nicolas931010/amylase_diversity_project/HPRC_AMY_Sequences/bundle_tree/outgroup
# mamba create -c conda-forge -c bioconda -n samtools samtools=1.17
conda activate samtools
rm outgroup_bundle?.fa
for SAMPLE in Altai Chagyrskaya Denisova Vindija; do
BAM=/global/scratch/p2p3/pl1_sudmant/human_diversity/ancient_genomes/neanderthals/${SAMPLE}/hg38/mapped.sorted.bam
echo '>'${SAMPLE} >> outgroup_bundle0.fa
samtools consensus -f fasta -r chr1:103456163-103571526 --min-MQ 20 --min-BQ 20 --line-len 1000000 \
${BAM} | tail -n 1 >>  outgroup_bundle0.fa
echo '>'${SAMPLE} >> outgroup_bundle1.fa
samtools consensus -f fasta -r chr1:103760698-103863980 --min-MQ 20 --min-BQ 20 --line-len 1000000 \
${BAM} | tail -n 1 >>  outgroup_bundle1.fa
done
cd ../fasta
cp bundle0.fa bundle0_with_outgroup.fa
cat ../outgroup/outgroup_bundle0.fa >> bundle0_with_outgroup.fa
cp bundle1.fa bundle1_with_outgroup.fa
cat ../outgroup/outgroup_bundle1.fa >> bundle1_with_outgroup.fag
```

## Multiple sequence alignment and tree construction (with kalign3 and iqtree)

```{bash eval=FALSE}
cd /global/scratch/users/nicolas931010/amylase_diversity_project/HPRC_AMY_Sequences/bundle_tree
## Set up kalign and iqtree
# mamba create -c bioconda -n kalign3 kalign3 iqtree
# mamba create -c bioconda -c conda-forge -n iqtree iqtree=2.2.2.3
## Run kalign and iqtree
for BUNDLE in 0 1; do
echo '#!/bin/bash
source ~/.bashrc
conda activate kalign3
cat fasta/bundle'${BUNDLE}'_with_outgroup.fa | kalign -o kalign/bundle'${BUNDLE}'_with_outgroup.afa &> kalign/bundle'${BUNDLE}'_with_outgroup.log' | \
sbatch \
--time=4320 \
--nodes=1 \
--cpus-per-task=10 \
--account=co_genomicdata \
--partition=savio4_htc \
--qos=savio_lowprio \
--output=logs/kalign/bundle${BUNDLE}-%j.log
done

## extract the part of bundle 1 that is upstream of the recombination hotspot as bundle 1a
# mamba create -c bioconda -c conda-forge -n seqkit seqkit
conda activate seqkit
seqkit subseq -r 1:66000 kalign/bundle1_with_outgroup.afa > kalign/bundle1a_with_outgroup.afa
conda deactivate

for BUNDLE in 0 1a; do
echo '#!/bin/bash
source ~/.bashrc
conda activate iqtree
iqtree -s kalign/bundle'${BUNDLE}'_with_outgroup.afa --prefix iqtree/bundle'${BUNDLE}'_with_outgroup -nt 10
iqtree -s kalign/bundle'${BUNDLE}'_with_outgroup.afa -te iqtree/bundle'${BUNDLE}'_with_outgroup.treefile --prefix iqtree/bundle'${BUNDLE}'_with_outgroup -nt 10 --date outgroup/bundle'${BUNDLE}'_date.txt --date-root -650000 -o "Altai,Chagyrskaya,Denisova,Vindija" --date-options "-l 0 -u 1" -undo' | \
sbatch \
--time=4320 \
--nodes=1 \
--cpus-per-task=10 \
--account=co_genomicdata \
--partition=savio4_htc \
--qos=savio_lowprio \
--output=logs/kalign/bundle${BUNDLE}-%j.log
done

#cat fasta/bundle'${BUNDLE}'.fa | kalign --type dna --nthreads 20 -f fasta > kalign/bundle'${BUNDLE}'.afa
#iqtree -te iqtree/bundle1a_with_outgroup.treefile --prefix iqtree/bundle1a_with_outgroup -nt 4 --date-root -650000 -o "Altai,Chagyrskaya,Denisova,Vindija" -s kalign/bundle1a_with_outgroup.afa --date outgroup/bundle1a_date.txt -undo --date-options "-l 0 -u 1" 

#--date-outlier 3
```

## Tree visualization

```{r eval=TRUE}
bundle_count_by_contig <- bundle_loc %>%
  group_by(chrom, name) %>%
  summarise(copy_number=n()) %>%
  pivot_wider(names_from = name, values_from = copy_number, values_fill=0, names_prefix = "copy_number_") %>%
  transmute(chrom = chrom, haplotype = str_c(copy_number_2, copy_number_3, copy_number_4, copy_number_5, copy_number_6))

bundle_info <- bundle_loc_filtered %>%
  transmute(chrom=chrom, start=start, end=end, name=name, 
            label=str_c(chrom, ".", start, "_", end) %>% str_replace_all("#", "_") %>% str_replace_all(":", "_"))
read_fasta <- function(x, range=NA){
  if (any(is.na(range))){
    dna <- read.dna(x, format = "fasta") 
  } else {
    dna <- read.dna(x, format = "fasta") %>% 
      .[,range]
  }
  return(dna)
}
nj_tree <- function(x, pairwise.deletion = TRUE){
  x %>%
    dist.dna(pairwise.deletion = TRUE) %>%
    nj() 
}
combine_tree_data <- function(x) {
  x %>%
    #midpoint.root() %>%
    as_tibble() %>%
    mutate(label = str_replace_all(label, "#", "_") %>% str_replace_all(":", "_")) %>%
    left_join(bundle_info, by="label") %>%
    left_join(hap_struc_table, by = c("chrom" = "contig")) %>%
    left_join(hap_group_table, by=c("hap_struc_d" = "label")) %>%
    left_join(gene_count) %>%
    as.treedata() 
}
plot_tree <- function(x, type, title){
  x %>%
    ggtree(layout=type) +
    geom_tippoint(mapping = aes(fill=group), size=3, color="black", shape = 21) +
    scale_fill_manual(values = c(group_color, "darkgrey")) +
    ggnewscale::new_scale_fill() +
    ggtitle(title) +
    theme_tree() +
    theme(legend.position = "top")
}
opposing_trees <- function(p1, p2, join_line_by){
  d1 <- p1$data
  d2 <- p2$data
  ## reverse x-axis and 
  ## set offset to make the tree on the right-hand side of the first tree
  d2$x <- max(d2$x) - d2$x + max(d1$x) + 0.0003
  pp <- p1 + 
    geom_tree(data=d2, layout='rectangular') +
    #geom_text2(data=d2, aes(subset=!isTip, label=node), hjust=-.3, size=2) +
    geom_tippoint(data=d2, mapping = aes(fill=group), size=3, color="black", shape = 21)
  dd <- bind_rows(d1, d2) %>% 
    filter(!is.na(label))
  pp + 
    geom_line(aes(x, y, group={{join_line_by}}, color=group), data=dd, linewidth = 0.2) +
    scale_color_manual(values = group_color) +
    scale_fill_manual(values = group_color)
}

```

#### All alignments

<details>
  <summary>Show figures</summary>

```{r fig.height=8, fig.width=15}
for (i in 0:1){
  dna <- read_fasta(str_c("kalign/bundle",i, "_with_outgroup.afa"))
  #spider::checkDNA(dna) %>%
  #  as.vector() %>%
  #  hist()
  seq_order <- spider::checkDNA(dna[,]) %>%
    as.vector() %>%
    order()
  ape::image.DNAbin(dna[seq_order,], show.labels = FALSE, xlab = str_c("bundle", i))
}
```

</details>

#### Ancestral state reconstruction with cafe5

```{r eval=FALSE, fig.width=6, fig.height=6}
for (bundle in c("0", "1a")){
  bundle_treedata <-str_c("iqtree/bundle", bundle, "_with_outgroup.timetree.nex") %>%
    read.nexus() %>%
    combine_tree_data() %>% 
    treeio::drop.tip(tip = c("Altai","Chagyrskaya", "Denisova", "Vindija"))
  #tips_to_drop <- bundle_treedata %>% as_tibble() %>% filter(!is.na(label)) %>% pull(label) %>% sample(50)
  rescaled_tree <- bundle_treedata %>%
    #treeio::drop.tip(tip=tips_to_drop) %>% 
    as_tibble() %>%
    mutate(branch.length=branch.length/1000)
  rescaled_tree %>%
    as.phylo() %>%
    write.tree(str_c("cafe/bundle", bundle, ".nwk"))
  cafe_table <- rescaled_tree %>%
    as_tibble() %>%
    dplyr::select(label, AMY1, AMY2A, AMY2B) %>%
    filter(!is.na(label)) %>%
    pivot_longer(cols = 2:4, names_to = "ID", values_to = "n") %>%
    pivot_wider(names_from = label, values_from = n, values_fill = 0) %>%
    mutate(DESCRIPTION=ID) %>%
    relocate(DESCRIPTION, ID) 
  cafe_table %>%
    write_tsv(str_c("cafe/bundle", bundle, "_copy_number.tsv"), col_names = TRUE)
  cafe_table %>%
    group_by(DESCRIPTION, ID) %>%
    group_walk(~write_tsv(.x, str_c("cafe/bundle", bundle, "_", .y$ID, "_copy_number.tsv")), .keep = TRUE)
}
```

All amy genes together

```{bash eval=FALSE}
#mamba create -c bioconda -c conda-forge -n cafe cafe
conda activate cafe
cd /global/scratch/users/nicolas931010/amylase_diversity_project/HPRC_AMY_Sequences/bundle_tree/
#mkdir -p cafe
for RUN in {1..5}; do
for BUNDLE in 0 1a; do
echo '#!/bin/bash
source ~/.bashrc
conda activate cafe
cd /global/scratch/users/nicolas931010/amylase_diversity_project/HPRC_AMY_Sequences/bundle_tree/
cafe5 --infile cafe/bundle'${BUNDLE}'_copy_number.tsv --tree cafe/bundle'${BUNDLE}'.nwk -k 3 --cores 20 -o cafe/bundle'${BUNDLE}'_results_run'${RUN} | \
sbatch \
--time=4320 \
--nodes=1 \
--cpus-per-task=20 \
--account=co_genomicdata \
--partition=savio4_htc \
--qos=savio_lowprio \
--output=cafe/bundle${BUNDLE}_run${RUN}.log
done
done
```

each gene separately

```{bash eval=FALSE}
for BUNDLE in 0 1a; do
for GENE in AMY1 AMY2A AMY2B; do
echo '#!/bin/bash
source ~/.bashrc
conda activate cafe
cd /global/scratch/users/nicolas931010/amylase_diversity_project/HPRC_AMY_Sequences/bundle_tree/
cafe5 --infile cafe/bundle'${BUNDLE}'_'${GENE}'_copy_number.tsv --tree cafe/bundle'${BUNDLE}'.nwk --cores 20 -o cafe/bundle'${BUNDLE}'_'${GENE}'_results' | \
sbatch \
--time=4320 \
--nodes=1 \
--cpus-per-task=20 \
--account=co_genomicdata \
--partition=savio4_htc \
--qos=savio_lowprio \
--output=cafe/bundle${BUNDLE}_${GENE}.log
done
done
```

estimated gene expansion rate

```{r eval=TRUE}
lambda_table_separate <- NULL
for (bundle in c("0", "1a")){
  for (gene in c("AMY1", "AMY2A", "AMY2B")){
    df <- read_tsv(str_c("../bundle_tree/cafe/bundle", bundle, "_", gene, "_results/Base_results.txt"), col_names = FALSE) %>%
      separate(X1, c("name", "value"), sep = ":", fill = "right") %>%
      mutate(value=parse_double(value)/1000)
    lambda_table_separate_tmp <- tibble(bundle=bundle, gene=gene, lambda=df[[2, 2]], likelihood=df[[1,2]])
    lambda_table_separate <- bind_rows(lambda_table_separate, lambda_table_separate_tmp)
  }
}
lambda_table_separate %>%
  dplyr::select(-likelihood) %>%
  pivot_wider(names_from = gene, values_from = lambda)

lambda_table_combined <- NULL
for (bundle in c("0", "1a")){
  for (run in 1:5){
    df <- read_tsv(str_c("../bundle_tree/cafe/bundle", bundle, "_results_run", run, "/Gamma_results.txt"), col_names = FALSE) %>%
      separate(X1, c("name", "value"), sep = ":", fill = "right") %>%
      mutate(value=parse_double(value)/1000)
    alpha <- read_tsv(str_c("../bundle_tree/cafe/bundle", bundle, "_results_run", run, "/Gamma_family_likelihoods.txt"), col_names = FALSE, skip = 1) %>%
      group_by(X1) %>%
      slice_max(X5) %>%
      ungroup() %>%
      transmute(gene=X1, lambda=X2*df[[2, 2]])
    lambda_table_combined_tmp <- alpha %>%
      mutate(bundle=bundle, run=run, likelihood=df[[1,2]])
    lambda_table_combined <- bind_rows(lambda_table_combined, lambda_table_combined_tmp)
  }
}
lambda_table_combined %>%
  pivot_wider(names_from = gene, values_from = lambda) %>%
  arrange(bundle, likelihood)
```

Ancestral state reconstruction visualization

```{r eval=TRUE, fig.width=6, fig.height=18, warning=FALSE, message=FALSE}
expSup <- function(w, digits=0) {
  sprintf(paste0("%.", digits, "f*x*10^%d"), w/10^floor(log10(abs(w))), floor(log10(abs(w))))
}

for (bundle in c("0", "1a")){
  for (gene in c("AMY1", "AMY2A", "AMY2B")){
    if(gene=="AMY1"){copy_number_break <-c(0,3,6,9)} else {copy_number_break <- c(0,1,2,3)}
    lambda <- read_tsv(str_c("../bundle_tree/cafe/bundle", bundle, "_", gene, "_results/Base_results.txt"), col_names = FALSE, show_col_types = FALSE) %>%
      separate(X1, c("name", "value"), sep = ":", fill = "right") %>%
      mutate(value=parse_double(value)/1000) %>%
      .[[2,2]]
    time_to_root <- str_c("cafe/bundle", bundle, "_", gene, "_results/Base_asr.tre") %>%
      read.nexus() %>% 
      castor::get_all_distances_to_root() 
    tree <- str_c("cafe/bundle", bundle, "_", gene, "_results/Base_asr.tre") %>%
      read.nexus() %>%
      as.treedata() %>%
      mutate(trait=str_extract(label, "(\\d+)$") %>% as.integer(), time_to_present=time_to_root-max(time_to_root))
    amy_ace_tree <- ggtree(tree, aes(color = trait), 
           layout = 'circular', 
           ladderize = TRUE, continuous = "color", size = 1.5) +
      geom_tiplab(aes(label=trait), hjust = -1, size = 5) +
      scico::scale_color_scico(palette = "bilbao", name = gene, begin = 0.05, limits=c(0, NA), midpoint = NA, direction = 1, breaks=copy_number_break, labels=copy_number_break) + 
      annotate(geom = "text", x=Inf, y=12, hjust = 0, vjust = -4, size=5, label = parse(text=str_c("lambda == ", expSup(lambda, 2)))) +
      theme(
        legend.position = c(0.05, 0.85),
        legend.text = element_text(size = 15),
        legend.title = element_blank())
    assign(str_c("p_", bundle, "_", gene), amy_ace_tree)
    ggsave(str_c("figures/bundle", bundle, "_", gene, "_ace_tree.pdf"), amy_ace_tree, width = 6, height = 6)
    amy_in_time <- tree %>%
      as_tibble() %>%
      ggplot(aes(x=time_to_present, y=trait)) +
      geom_point()+
      geom_smooth(method="lm") +
      theme_cowplot() +
      theme(panel.border = element_rect(color="black", linewidth = 0.5))
    assign(str_c("time_", bundle, "_", gene), amy_in_time)
    lm(trait~time_to_present, tree%>%as_tibble) %>% summary() %>% .$coefficients %>% knitr::kable()
  }
}
amy_ace_plot_bundle0 <- plot_grid(p_0_AMY1, p_0_AMY2A, p_0_AMY2B, ncol = 1)
amy_ace_plot_bundle1a <- plot_grid(p_1a_AMY1, p_1a_AMY2A, p_1a_AMY2B, ncol = 1)
amy_ace_plot_bundle0
amy_ace_plot_bundle1a
time_plot_bundle0 <- plot_grid(time_0_AMY1, time_0_AMY2A, time_0_AMY2B, ncol = 1)
time_plot_bundle1a <- plot_grid(time_1a_AMY1, time_1a_AMY2A, time_1a_AMY2B, ncol = 1)
time_plot_bundle0
time_plot_bundle1a
```

```{r eval=FALSE}
ggsave("figures/bundle0_ace_tree.pdf", amy_ace_plot_bundle0, width = 6, height = 18)
ggsave("figures/bundle1a_ace_tree.pdf", amy_ace_plot_bundle1a, width = 6, height = 18)
```

#### Rectangular

Bundle 0

```{r fig.height=12, fig.width=10}
## with haplotype names
bundle0_tree <- str_c("iqtree/bundle", 0, "_with_outgroup.timetree.nwk") %>%
  read.tree() %>%
  combine_tree_data() %>% 
  mutate(group=ifelse(is.na(group), "outgroup", group)) %>%
  plot_tree("rectangular", "bundle0")
bundle0_tree +
  geom_tiplab(aes(label=ifelse(is.na(chrom), label, chrom)), align = TRUE, size = 3, hjust = 0, offset = 0.00002) +
  xlim_tree(0.0015)
## with amy copy numbers
bundle0_histo_dots <- bundle_info %>%
  filter(name=="0") %>%
  left_join(hap_struc_table, by=c("chrom"="contig")) %>%
  dplyr::select(label, hap_struc_d) %>%
  left_join(histo_dots, by = "hap_struc_d") %>%
  bind_rows(tibble(label=rep(c("Altai", "Chagyrskaya", "Denisova", "Vindija"), each=3),
                   name=rep(c("AMY1", "AMY2A", "AMY2B"),times=4),
                   index=rep(1:3, times=4)))
bundle0_histo_dots_plot <- bundle0_histo_dots %>%
  ggplot(aes(x=index, y=label, fill=name, shape=name)) +
  geom_point(size=3) +
  scale_shape_manual(values = 22:25) +
  scale_fill_manual(values = amy_color) +
  ggnewscale::new_scale_fill() +
  theme_void() +
  theme(axis.text.y = element_blank())
bundle0_tree_with_gene_count <- bundle0_histo_dots_plot %>%
  aplot::insert_left(bundle0_tree +
                       geom_tiplab(mapping = aes(label=ifelse(is.na(amy_copy_number), "1-1-0-1", amy_copy_number)), align = TRUE, size = 3, hjust = -0.1, offset = 0.00002) +
                       xlim_tree(0.00068), width = 2.5)
bundle0_tree_with_gene_count
```

```{r eval=FALSE}
ggsave("figures/bundle0_tree_with_gene_count.pdf", bundle0_tree_with_gene_count, height = 12, width = 10, units = "in")
## For Joana's PCA
bundle_info %>%
  filter(name=="0") %>%
  left_join(hap_struc_table, by=c("chrom"="contig")) %>%
  dplyr::select(chrom, label, hap_struc_d) %>%
  left_join(histo_dots, by = "hap_struc_d") %>%
  count(chrom, name) %>%
  pivot_wider(names_from = name, values_from = n, values_fill = 0) %>%
  dplyr::select(-AMY2Ap) %>%
  write_tsv("gene_count_by_haplotype.tsv")
```

Bundle 1a

```{r fig.height=12, fig.width=10}
## with haplotype names
bundle1a_tree <- str_c("iqtree/bundle", "1a", "_with_outgroup.timetree.nwk") %>%
  read.tree() %>%
  combine_tree_data() %>% 
  mutate(group=ifelse(is.na(group), "outgroup", group)) %>%
  plot_tree("rectangular", "bundle1a")
bundle1a_tree +
  geom_tiplab(aes(label=ifelse(is.na(chrom), label, chrom)), align = TRUE, size = 3, hjust = 0, offset = 0.00002) +
  xlim_tree(0.0015)
## with amy copy numbers
bundle1a_histo_dots <- bundle_info %>%
  filter(name=="1") %>%
  left_join(hap_struc_table, by=c("chrom"="contig")) %>%
  dplyr::select(label, hap_struc_d) %>%
  left_join(histo_dots, by = "hap_struc_d") %>%
  bind_rows(tibble(label=rep(c("Altai", "Chagyrskaya", "Denisova", "Vindija"), each=3),
                   name=rep(c("AMY1", "AMY2A", "AMY2B"),times=4),
                   index=rep(1:3, times=4)))
bundle1a_histo_dots_plot <- bundle1a_histo_dots %>%
  ggplot(aes(x=index, y=label, fill=name, shape=name)) +
  geom_point(size=3) +
  scale_shape_manual(values = 22:25) +
  scale_fill_manual(values = amy_color) +
  ggnewscale::new_scale_fill() +
  theme_void() +
  theme(axis.text.y = element_blank())
bundle1a_tree_with_gene_count <- bundle1a_histo_dots_plot %>%
  aplot::insert_left(bundle1a_tree +
                       geom_tiplab(mapping = aes(label=ifelse(is.na(amy_copy_number), "1-1-0-1", amy_copy_number)), align = TRUE, size = 3, hjust = -0.1, offset = 0.00002) +
                       xlim_tree(0.00083), width = 2.5)
bundle1a_tree_with_gene_count
```

```{r eval=FALSE}
ggsave("figures/bundle1a_tree_with_gene_count.pdf", bundle1a_tree_with_gene_count, height = 12, width = 10, units = "in")
```

```{r eval=FALSE, fig.height=8, fig.width=15}
dna <- read_fasta(str_c("kalign/bundle",0, "_with_outgroup.afa"))
missing_data_location <- dna %>% as.character() %>% apply(2, function(x){sum(x=="-")}) %>% magrittr::divide_by(dim(dna)[1]) %>% magrittr::is_greater_than(0.1) %>% which()
common_snp_location <- dna[,] %>% as.character() %>% apply(2, function(x){table(x) %>% sort(decreasing = TRUE) %>% .[1] %>% magrittr::divide_by(length(x))}) %>% magrittr::is_less_than(0.9) %>% which()
common_snp <- dna[, setdiff(common_snp_location, missing_data_location)]
common_snp <- updateLabel(common_snp, labels(common_snp), labels(common_snp) %>% str_replace_all("#", "_") %>% str_replace_all(":", "_"))
msaplot(bundle0_tree, common_snp, offset=0.0001, width=3, window = NULL, color = c("black", "red", "yellow", "green", "white", "blue"), bg_line=FALSE, height=1)
```


#### IQTree vs. Neighbor-joining  

bundle 0

```{r fig.width=8, fig.height=10}
for (bundle in c("0", "1a")){
  p1 <- str_c("iqtree/bundle", bundle, "_with_outgroup.timetree.nwk") %>%
    read.tree() %>%
    combine_tree_data() %>%
    plot_tree("rectangular", str_c("bundle", bundle, ": iqtree vs. neighbor-joining")) +
    geom_text2(aes(subset=!isTip, label=node), hjust=-.3, size=2)
  p2 <- str_c("kalign/bundle", bundle, "_with_outgroup.afa") %>%
    read_fasta() %>%
    nj_tree() %>%
    midpoint.root() %>%
    combine_tree_data() %>%
    ggtree()
  assign(str_c("p1_bundle", bundle), p1)
  assign(str_c("p2_bundle", bundle), p2)
}
opposing_trees(p1_bundle0 %>% ggtree::rotate(149) %>% ggtree::rotate(150), 
               p2_bundle0 %>% ggtree::rotate(111) %>% ggtree::rotate(144), 
               label)
opposing_trees(p1_bundle1a, 
               p2_bundle1a %>% ggtree::rotate(110) %>% ggtree::rotate(111)%>% ggtree::rotate(112)  %>% ggtree::rotate(137), 
               label)
```

#### Trees with first vs. second half of bundle 1

```{r eval=FALSE, fig.width=8, fig.height=8}
bundle1_window1 <- str_c("kalign/bundle", 1, ".afa") %>%
  read_fasta(00001:70018) %>%
  nj_tree() %>%
  combine_tree_data() %>%
  plot_tree("rectangular", str_c("The first vs. second half of bundle 1")) +
  geom_text2(aes(subset=!isTip, label=node), hjust=-.3, size=2)
bundle1_window2 <- str_c("kalign/bundle", 1, ".afa") %>%
  read_fasta(70019:103444) %>%
  nj_tree() %>%
  combine_tree_data() %>%
  ggtree()+
  geom_text2(aes(subset=!isTip, label=node), hjust=-.3, size=2)
bundle1_opposing_trees <- opposing_trees(bundle1_window1 %>% ggtree::rotate(94) %>% ggtree::rotate(120) %>% ggtree::rotate(125) %>% ggtree::rotate(128), 
                                         bundle1_window2 %>% ggtree::rotate(92) %>% ggtree::rotate(93) %>% ggtree::rotate(134), label)
bundle1_opposing_trees
ggsave("figures/bundle1_opposing_trees.pdf", bundle1_opposing_trees, width = 8, height = 8, unit="in")
```

#### Bundle 0 vs. 1a

```{r fig.width=10, fig.height=12}
bundle0 <- read.tree("iqtree/bundle0_with_outgroup.timetree.nwk") %>%
  combine_tree_data() %>%  
  mutate(chrom=ifelse(is.na(chrom)&!is.na(label), label, chrom)) %>%
  plot_tree("rectangular", str_c("Bundle 0 vs. 1a")) 
bundle1a <- read.tree("iqtree/bundle1a_with_outgroup.timetree.nwk") %>%
  combine_tree_data() %>%  
  mutate(chrom=ifelse(is.na(chrom)&!is.na(label), label, chrom)) %>%
  ggtree() %>% 
  ggtree::rotate(110) %>% ggtree::rotate(107) 
bundle0_vs_bundle1_opposing_trees <- opposing_trees(bundle0 #+ geom_text2(aes(subset=!isTip, label=node), hjust=-.3, size=2)
                                                    ,bundle1a , 
                                                    chrom)
bundle0_vs_bundle1_opposing_trees
ggsave("figures/bundle0_vs_bundle1_opposing_trees.pdf", bundle0_vs_bundle1_opposing_trees, width = 10, height = 12, unit="in")
```

#### Plot with 3 trees

```{r eval=FALSE,fig.width=15, fig.height=8}
p1 <- bundle0_window2 %>% ggtree::rotate(92) %>% ggtree::rotate(93) %>% ggtree::rotate(129) %>% ggtree::rotate(133) %>% ggtree::rotate(155) %>% ggtree::rotate(164)
p2 <- bundle1_window1 %>% ggtree::rotate(94) %>% ggtree::rotate(120) %>% ggtree::rotate(125) %>% ggtree::rotate(128)
p3 <- bundle1_window2 %>% ggtree::rotate(92) %>% ggtree::rotate(93) %>% ggtree::rotate(134)
d1 <- p1$data
d2 <- p2$data
d3 <- p3$data
## reverse x-axis and 
## set offset to make the tree on the right-hand side of the first tree
d2$x <- max(d2$x) - d2$x + max(d1$x) + 0.00015
d3$x <- max(d3$x) - d3$x + max(d2$x) + 0.00015

pp <- p1 + 
  geom_tree(data=d2, layout='rectangular') +
  geom_tree(data=d3, layout='rectangular') +
  #geom_text2(data=d2, aes(subset=!isTip, label=node), hjust=-.3, size=2) +
  geom_tippoint(data=d2, mapping = aes(fill=group), size=3, color="black", shape = 21) +
  geom_tippoint(data=d3, mapping = aes(fill=group), size=3, color="black", shape = 21) 
dd <- bind_rows(d1, d2, d3) %>% 
  filter(!is.na(label))
pp + 
  geom_line(aes(x, y, group=chrom, color=group), data=dd, size = 0.3) +
  scale_color_manual(values = group_color) +
  scale_fill_manual(values = group_color) +
  theme(legend.position = "none",
        title = element_blank())
```


## Diversity

```{r eval=FALSE}
nuc_div_in_windows <- function(dna, window_size, length_to_trim){
  ## Trim the beginning and the end
  length <- dim(dna)[2]
  dna <- dna[, -c(0:length_to_trim, (length-length_to_trim+1):(length+1))]
  length <- dim(dna)[2]
  n_windows <- (length/window_size) %>% ceiling()
  windows <- lapply(0:(n_windows-2), function(x){c(1+x*window_size, window_size+x*window_size)}) 
  windows[[n_windows]]  <- c(1+(n_windows-1)*window_size, length)
  pos <- lapply(windows, mean) %>%
    unlist() %>%
    floor() %>%
    magrittr::add(length_to_trim)
  size <- lapply(windows, function(x){x[2]-x[1]}) %>%
    unlist()
  lapply(windows, function(x){dna[,x[1]:x[2]]}) %>%
    lapply(function(x){nuc.div(x, pairwise.deletion = FALSE)}) %>%
    unlist() %>%
    tibble(pos=pos, size=size, nuc_div=.)
}
nuc_div_table <- NULL
for(i in as.character(0:5)){
  nuc_div_table_tmp <- read_fasta(str_c("kalign/bundle", i, ".afa")) %>%
    nuc_div_in_windows(5000, 500) %>%
    mutate(bundle=i)
  nuc_div_table <- bind_rows(nuc_div_table, nuc_div_table_tmp)
}
pi_in_windows <- nuc_div_table %>%
  ggplot(aes(x=pos/1000, y=nuc_div)) +
  geom_line() +
  geom_point() +
  facet_wrap(~bundle, scales = "free_x") +
  cowplot::theme_cowplot() +
  labs(x="position in kb", y="nucleotide diversity")+
  theme(panel.border = element_rect(color="black", size=1))
pi_in_windows
ggsave("../bundle_tree/figures/pi_in_windows.pdf", pi_in_windows, height = 6, width = 8, units = "in")
```

Note `pairwise.deletion` makes a difference in some cases. Need to think more about it.

```{r eval=FALSE, fig.width=12, fig.height=8}
contig_group_table <- hap_struc_table %>%
  left_join(hap_group_table, by=c("hap_struc_d"="label"))
pairwise_distance=NULL
for (i in 0:5){
  pairwise_distance_tmp <- str_c("kalign/bundle", i, ".afa") %>%
    read_fasta() %>%
    dist.dna(model = "raw", pairwise.deletion = TRUE) %>%
    broom::tidy() %>%
    mutate(bundle = str_c("bundle", i),
           ind1=str_extract(item1, "^.*?(?=[#\\.])"),
           ind2=str_extract(item2, "^.*?(?=[#\\.])"),
           contig1=str_extract(item1, ".*(?=\\.[^.]*$)"),
           contig2=str_extract(item2, ".*(?=\\.[^.]*$)")) %>%
    left_join(contig_group_table, by = c("contig1"="contig")) %>%
    left_join(contig_group_table, by = c("contig2"="contig"), suffix = c("1", "2")) %>%
    mutate(type = case_when(contig1==contig2 ~ "same contig",
                            hap_struc_d1==hap_struc_d2 ~ "same structure",
                            group1==group2 ~ "same group", 
                            TRUE ~ "others"))
  pairwise_distance <- bind_rows(pairwise_distance, pairwise_distance_tmp)
}
set.seed(42)
pairwise_distance_plot <- pairwise_distance %>%
  mutate(group=ifelse(group1=="6" | group2=="6", "comparisons involving group 6", "comparisons not involving group 6")) %>%
  bind_rows(pairwise_distance %>% mutate(group="all")) %>%
  filter(! bundle %in% c("bundle4", "bundle5")) %>%
  mutate(type=fct_relevel(type, c("same contig", "same structure", "same group", "others"))) %>%
  ggplot(aes(x=fct_rev(type), y=distance)) +
  geom_boxplot(outlier.alpha = 0) +
  #geom_jitter(size = 0.2, width = 0.1) +
  scattermore::geom_scattermore(pointsize = 1, position=position_jitter(height=0, width = 0.2)) +
  labs(y="pairwise distance") +
  coord_flip() +
  facet_grid(bundle~group, scales = "free_x") +
  cowplot::theme_cowplot() +
  theme(panel.border = element_rect(color="black", size=1),
        axis.title.y = element_blank())
pairwise_distance_plot
ggsave("figures/pairwise_distance_plot.pdf", pairwise_distance_plot, width = 12, height = 8, units = "in")
```
