---
title: "Untitled"
output: html_document
date: '2022-07-21'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Code to output "raw" read depth based genotypes using the raw depth and the normalization factor

```{r}
#library(ggcorrplot)
library(tidyverse)
library(PNWColors)
exclude = c("1KG.HG00403.CHS.EA", "1KG.HG00662.CHS.EA", "1KG.HG01773.IBS.WEA", "1KG.HG01857.KHV.EA", 
            "1KG.NA18644.CHB.EA", "1KG.NA18916.YRI.AFR", "1KG.NA18985.JPT.EA", "1KG.NA19159.YRI.AFR",
            "1KG.NA19201.YRI.AFR", "1KG.NA19204.YRI.AFR", "1KG.NA20892.GIH.SA") 

exclude = c(exclude,"HGDP.HGDP00351.Burusho.SA",
                    "1KG_trio.HG00405.CHS.EA",
                    "1KG_trio.HG00664.CHS.EA")

ancient_qc = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/info/ancients_passed_QC.tsv", header=T,sep="\t")

outdir="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/read_depth_genotyping/genotyping/output"
fn = paste(outdir,"/genome_summary_stats.tsv",sep="")
stats = read.table(fn,header=T,sep="\t") %>%
        filter(cvg_med>0) %>%
        filter(!(sample %in% exclude)) %>%
        separate(sample,c("group","ID","y","z"),sep="\\.",remove=FALSE) %>%
        dplyr::select(-c(y,z)) %>%
        
  #separate(name,into = c("source","ID","pop","region"),sep="\\.")
#samples w/ median 0

t_wnd_raw = data.table::fread("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/input/combined/amylase_1000_200.1000.200.tsv.gz",header=F,sep="\t",col.names = c("contig","start","end","cvg","sample")) %>%
  inner_join(stats,on="sample") %>%
  mutate(cp = 2*cvg/cvg_mu) %>%
  #mutate(cp = 2*cvg/cvg_med) %>%
  separate(sample,c("source","sample_id",'p1','p2'),remove=FALSE,sep="\\.") %>%
  filter(start>0)

# sds = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/data/superdups_headered.txt",header=T,sep="\t") %>%
#   filter(chrom=="chr1") %>%
#   filter(chromEnd<103820315,chromStart>103499858) %>%
#   mutate(start_1 = chromStart, 
#          end_1 = chromEnd,
#          start_2 = otherStart,
#          end_2 = otherEnd) %>% 
#   dplyr::select(start_1,end_1,start_2,end_2,strand,fracMatch) %>%
#   group_by(fracMatch,strand) %>%
#   summarize(start_1 = first(start_1),
#             start_2 = first(start_2),
#             end_1 = first(end_1),
#             end_2 = first(end_2))%>%
#   ungroup() %>%
#   #filter(fracMatch>0.93) %>%
#   mutate(ix = row_number()) %>%
#   arrange(fracMatch)

# usher_blocks = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/data/amylase_gene_indices_hg38.tsv",header=T) %>%
#   filter(Type!="hg38_Gene") %>%
#   mutate(ix=row_number())
# fn_info="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/info/sample_info.tsv"
# sample_info = read.table(fn_info,sep="\t",header=T) %>% 
#   mutate(sample_id = sample) %>%
#   dplyr::select(-c(sample))

# cp_sd_cutoff = 2.5
# t_wnd_cvg = t_wnd_raw %>% 
#   inner_join(sample_info,by="sample_id") %>%
#   filter(cp_sd<cp_sd_cutoff) %>%
#   filter(!isAncient)


#usher_blocks
t_wnd_raw %>% 
  dplyr::select(source) %>% 
  group_by(source) %>%
  summarize(n=n()/1244)

stats %>% 
  dplyr::select(group)%>%
  group_by(group) %>%
  summarize(n=n())

t_wnd_raw %>% filter(p2=="Europe")

```


```{r}
modern_colors = pnw_palette("Bay",7)[1:8]
#modern_colors = pnw_palette("Bay",7)[1:5]
ancient_colors = pnw_palette("Sunset2",7)[1:4]

colors = c(ancient_colors,modern_colors)


genes = data.frame(names=c("RNPC3", "AMY2B", "AMY2A","AMY1A","AMY1B","AMY1C","AMY2Ap"),
                   start=c(103525691,103571100,103617427,103655760,103687415,103749898,103713720),
                   end=c(103553343,103579530,103625780,103664554,103696210,103758692,  103719905))

# genotype_blocks = data.frame(name=c("AMY2B", "AMY2A","AMY1A","AMY1B","AMY1C","prox"),
#                              start=c(103564100,103612500,103635500,103698415,103729898,103762000),
#                              end=  c(103588530,103617000,103655000,103716210,103743692,103795000))
#   
# #these work quite well, and it turns out they are INSIDE the usher ones
# genotype_blocks = data.frame(name=c("AMY2B", "AMY2A","AMY1A","AMY1B","AMY1C","prox"),
#                              start=c(103564100,103612500,103639500,103700415,103732500,103762000),
#                              end=  c(103588530,103617000,103655000,103715210,103746000,103795000))

#Usher coords for Amy 1

genotype_blocks = data.frame(name=c("AMY2B", "AMY2A","AMY1A","AMY1B","AMY1C","AMY2Ap"),
                             start=c(103564100,103612500,103638544,103684142,103732686,103713720),
                             end=  c(103588530,103617000,103667876,103712474,103762027,103730686))

control_blocks = data.frame(name=c("left_ctrl","right_ctrl"),
                            start = c(103550201,      103765027),
                            end =   c(103550201+10000,103762027+25000))

genotype_blocks =rbind(genotype_blocks,control_blocks)
outdir="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/read_depth_genotyping/genotyping/output_figures"
fn_out = paste(outdir,"/amylase_distribution.pdf",sep="")
pdf(fn_out,width=8,height=8)

g=ggplot(t_wnd_raw)
g+geom_line(aes(x=start,y=cp,group=sample,color=p2),size=.1,alpha=.5)+
  theme_bw()+
  facet_wrap(~p2,ncol=1,strip.position = "right")+
  scale_y_continuous(lim=c(0,7),breaks=seq(0,10,1))+
  theme(legend.position="None")+
  theme(strip.background = element_blank())+
  scale_color_manual(values=colors)+
  geom_segment(aes(x=start,xend=end,y=0.5,yend=0.5),size=0.5,data=genes)+
  geom_segment(aes(x=start,xend=end,y=0,yend=0),size=0.5,data=genotype_blocks,color='blue')+
  geom_vline(aes(xintercept=start),size=0.15,data=genotype_blocks,color='blue',linetype='dotted')+
  geom_vline(aes(xintercept=end),size=0.15,data=genotype_blocks,color='blue',linetype='dotted')

dev.off()

```



```{r}

#pop_info = t %>% select(sample,popgrouping,popgrouping2,ageaverage,isAncient) %>% unique

t_gt = list()
for (i in seq(dim(genotype_blocks)[1])){
  curr_start = genotype_blocks$start[i]
  curr_end = genotype_blocks$end[i]
  locus = genotype_blocks$name[i]
  size = curr_end-curr_start
  t_sub_gt = t_wnd_raw %>% 
    filter(start>curr_start,end<curr_end) %>%
    group_by(sample) %>%
    summarize(gt = mean(cp)) %>%
    #summarize(gt = median(cp)) %>%
    mutate(locus=locus)
  t_gt[[i]]=t_sub_gt
}
t_gt = do.call(rbind, t_gt) %>%
        pivot_wider(values_from = "gt",names_from="locus") %>%
        mutate(amy1_sum = AMY1A + AMY1B + AMY1C) %>%
        pivot_longer(cols=-c("sample"),values_to="gt") %>%
        separate(sample,c("dataset","ID","p1","p2"),sep="\\.",remove=FALSE)
  #inner_join(t_stats,on="sample") 

#### MODIFY THE SGDP AND HGDP BY A CONSTANT FACTOR SO THEY FIT INTEGERS

t_gt_adj = t_gt %>% mutate(gt=ifelse(dataset=="SGDP",gt*1.08,gt)) %>%
         mutate(gt=ifelse(dataset=="HGDP",gt*0.97,gt))



outdir="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/read_depth_genotyping/genotyping/output"
fn_out = paste(outdir,"/genotypes.raw.tsv",sep="")
write.table(t_gt_adj,fn_out,sep="\t",row.names = FALSE,quote=FALSE)

length(unique(t_gt$sample))
```
```{r}
#find bad samples
#g=ggplot(t_gt %>% filter(name %in% c("left_ctrl","right_ctrl")))# %>% filter(!grepl("StoneAgeAncient",dataset)))
g=ggplot(t_gt %>% filter(name %in% c("left_ctrl","right_ctrl")) %>% filter(!grepl("StoneAgeAncient",dataset)))
g+geom_histogram(aes(x=gt,fill=p2))+
  facet_wrap(~name)+
  ylim(c(0,100))

t_gt %>% filter(name=="left_ctrl", gt>4)
t_gt %>% filter(name=="right_ctrl", gt<1)

#HGDP.HGDP00351.Burusho.SA	
#1KG_trio.HG00405.CHS.EA	
#1KG_trio.HG00664.CHS.EA	
```

```{r}

outdir="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/read_depth_genotyping/genotyping/output_figures"

fn_out = paste(outdir,"/genotypes_amy1_sum.pdf",sep="")
pdf(fn_out,width=3,height=2)
# g=ggplot(t_gt %>% 
#            filter(name == "amy1_sum") %>% 
#            mutate(gt=ifelse(dataset=="SGDP",gt*1.08,gt)) %>%
#            mutate(gt=ifelse(dataset=="HGDP",gt*0.97,gt)))
g=ggplot(t_gt %>% 
           filter(name == "amy1_sum"))
g+geom_histogram(aes(x=gt),binwidth=0.05)+
  theme_bw(base_size=8)+
  facet_wrap(~dataset,scales="free_y",ncol=1)+
  scale_x_continuous(breaks=seq(0,18,1))+
  coord_cartesian(xlim=c(0,20))+
  theme(legend.position = "None",
        strip.background = element_blank(),
        strip.text = element_text(size=4))
dev.off()
# unique(t_gt$p1)
# t_test = t_gt %>% 
#           filter(name == "amy1_sum") %>% 
#           filter(dataset%in% c("1KG","StoneAgeAncient")) %>%
#           filter((dataset =="StoneAgeAncient")|(p1 %in% c("GBR","FIN","TSI")))
# g=ggplot(t_test)
# g+geom_boxplot(aes(x=reorder(p2,gt),y=gt))+
#   theme_bw(base_size=8)

# g=ggplot(t_test)
# g+geom_point(aes(x=reorder(p2,gt),y=gt),position=position_jitter(width=.1,height=0))+
#   theme_bw(base_size=8)
# 
# 
# g=ggplot(t_test)
# g+stat_summary(aes(x=reorder(p1,gt),y=gt))+
#   theme_bw(base_size=8)


fn_out = paste(outdir,"/genotypes_amy2A.pdf",sep="")
pdf(fn_out,width=2 ,height=2)
g=ggplot(t_gt %>% filter(name == "AMY2A"))
g+geom_histogram(aes(x=gt),binwidth=0.05)+
  theme_bw(base_size=8)+
  facet_wrap(~dataset,scales="free_y",ncol=1)+
  scale_x_continuous(breaks=seq(0,8,1))+
  theme(legend.position = "None",
        strip.background = element_blank(),
        strip.text = element_text(size=4))

  #facet_grid(isAncient~.,scales="free_y")
dev.off()
# 
fn_out = paste(outdir,"/genotypes_amy2B.pdf",sep="")
pdf(fn_out,width=2,height=2)
g=ggplot(t_gt %>% filter(name == "AMY2B"))

g+geom_histogram(aes(x=gt),binwidth=0.05)+
  theme_bw(base_size=8)+
  facet_wrap(~dataset,scales="free_y",ncol=1)+
  scale_x_continuous(breaks=seq(0,8,1))+
  theme(legend.position = "None")
  theme(legend.position = "None",
        strip.background = element_blank(),
        strip.text = element_text(size=4))

dev.off()


t_combined = t_gt %>% 
  filter(name %in% c("AMY2A","amy1_sum","AMY2B")) %>%
  pivot_wider(names_from=name,values_from=gt)

fn_out = paste(outdir,"/genotypes_amy2B_vs_amy1.pdf",sep="")
pdf(fn_out,width=5,height=3)
g=ggplot(t_combined)
g+geom_point(aes(x=amy1_sum,y=AMY2B))+
  theme_bw(base_size=8)+
  facet_wrap(~dataset,scales="free_y",ncol=1)+
  theme(legend.position = "None",
        strip.background = element_blank(),
        strip.text = element_text(size=4))

  #$facet_grid(isAncient~.,scales="free_y")
dev.off()

fn_out = paste(outdir,"/genotypes_amy2A_vs_amy1.pdf",sep="")
pdf(fn_out,width=5,height=3)
g=ggplot(t_combined)
g+geom_point(aes(x=amy1_sum,y=AMY2A))+
  theme_bw(base_size=8)+
  facet_wrap(~dataset,scales="free_y",ncol=1)+
  theme(legend.position = "None",
        strip.background = element_blank(),
        strip.text = element_text(size=4))

dev.off()

fn_out = paste(outdir,"/genotypes_amy2A_vs_amy2B.pdf",sep="")
pdf(fn_out,width=5,height=3)
g=ggplot(t_combined)
g+geom_point(aes(x=AMY2B,y=AMY2A))+
  theme_bw(base_size=8)+
  facet_wrap(~dataset,scales="free_y",ncol=1)+
  theme(legend.position = "None",
        strip.background = element_blank(),
        strip.text = element_text(size=4))

dev.off()







# 
# fn_out = paste(outdir,"/genotypes_amy1_sum_dist.pdf",sep="")
# pdf(fn_out,width=2.5,height=3)
# g=ggplot(t_gt %>% filter(name == "amy1_sum"))
# 
# g+geom_point(aes(y=gt,x=reorder(popgrouping2,gt),color=popgrouping2),alpha=.25,size=0.1,position=position_jitter(width=.1,height=0))+
#   geom_boxplot(aes(y=gt,x=reorder(popgrouping2,gt),color=popgrouping2,fill=popgrouping2,alpha=isAncient),outlier.shape=NA)+
#   theme_bw(base_size=8)+
#   theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5))+
#   scale_color_manual(values = c(ancient_colors,modern_colors))+
#   scale_fill_manual(values = c(ancient_colors,modern_colors))+
#   theme(legend.position="None")+
#   scale_alpha_manual(values=c(0,0.5))
# 
# dev.off()

outdir="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/output/d4/genotypes"

#fn_out = paste(outdir,"/genotypes_amy1_sum.pdf",sep="")
#pdf(fn_out,width=2.5,height=3)
# g=ggplot(t_gt %>% filter(name == "prox"))
# g+geom_histogram(aes(x=gt),binwidth=0.05)+
#   theme_bw(base_size=8)+
#   geom_vline(xintercept=1.5)+
#   geom_vline(xintercept=2.1)
#dev.off()

off_samples = (t_gt %>% 
               filter(name == "prox") %>%
               filter((gt<1.5)|(gt>2.1)))$sample

  
```
```{r}

s=103780000
e=103799858

simple_repeats = read.table("/Users/petersudmant/tmp/amylase/input/repeats/simpleRepeat.txt.gz",header=F,sep="\t") %>%
                  mutate(chr=V2,start=V3,end=V4) %>%
                  dplyr::select(chr,start,end) %>%
                  filter(chr=="chr1",start>s,end<e)
trf_alma = read.table("/Volumes/GoogleDrive/Shared\ drives/sudmantlab/projects/ancient_DNA/masking/hg38.trf.bed",header=F,sep="\t") %>%
                  mutate(chr=V1,start=V2,end=V3) %>%
                  dplyr::select(chr,start,end) %>%
                  filter(chr=="chr1",start>s,end<e)

wm = read.table("/Users/petersudmant/tmp/amylase/input/repeats/windowmaskerSdust.txt.gz",header=F,sep="\t") %>%
                  mutate(chr=V2,start=V3,end=V4) %>%
                  dplyr::select(chr,start,end) %>%
                  filter(chr=="chr1",start>s,end<e)


x=t_wnd_raw %>% 
  filter(grepl("StoneAgeAncient",sample)) %>%
  filter(start>103670000,end<103799858)


x_sum = x %>% group_by(sample) %>%summarize(mu = mean(cp))
g=ggplot(x)
g+geom_line(aes(x=start,y=cp,group=sample),size=.1,alpha=1) +
  geom_segment(aes(x=start,xend=end,y=-1,yend=-1),data=simple_repeats,color='blue',size=1)+
  geom_segment(aes(x=start,xend=end,y=-2,yend=-2),data=wm)+
  geom_segment(aes(x=start,xend=end,y=-3,yend=-3),color='red',size=1,data=trf_alma)+
  scale_y_continuous(lim=c(-3,15))+
  geom_hline(aes(yintercept=2),color='red')

#g=ggplot(x_sum)
#g+geom_histogram(aes(x=mu),binwidth=0.01)
```
```{r}

outdir="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/output/d4/heatmaps"

X = t_wnd_raw %>% 
  dplyr::select(sample,start,cp) %>%
  pivot_wider(values_from = "cp",
              names_from = "start")
X_mat = as.matrix(X %>% dplyr::select(-c(sample)))
c = cor(X_mat)^2
df_c = as.data.frame.table(c) %>% mutate(start = as.numeric(Var1), start_2 = as.numeric(Var2))



fn_out = paste(outdir,"/corr_plot_locus_all.pdf",sep="")
pdf(fn_out,width=4,height=4)
ggcorrplot(c,outline.color = "NA")+
#ggplot(df_c)+
  #geom_tile(aes(x=start,start_2,fill=Freq))+
  #theme_bw(base_size=8)+
  scale_fill_viridis_c()+
  ggtitle("All")+
  theme(legend.position="bottom")
dev.off()

for (p in unique(t_wnd_raw$p2)){
  X = t_wnd_raw %>% 
    filter(p2==p) %>%
    dplyr::select(sample,start,cp) %>%
    pivot_wider(values_from = "cp",
                names_from = "start")
  X_mat = as.matrix(X %>% dplyr::select(-c(sample)))
  c = cor(X_mat)^2

  fn_out = paste(outdir,"/corr_plot_locus_",p,".pdf",sep="")
  pdf(fn_out,width=4,height=4)
  g= ggcorrplot(c,outline.color = "NA")+
    theme_bw(base_size=8)+
    theme(legend.position="None")+
    scale_fill_viridis_c(begin = 0)+
    ggtitle(p)
  print(g)
  dev.off()
    

}




```


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
