---
title: "Untitled"
output: html_document
date: "2023-02-21"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r}
library(tidyverse)
#install.packages("devtools") 
#devtools::install_github("jakelawlor/PNWColors") 
library(PNWColors)
library(ggplot2)
theme_set(theme_bw())
library(stringr)
library(dplyr)
library(tidyr)
library(Rcpp)

outdir="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/read_depth_genotyping/genotyping/output_figures"

t_gt=read.table("./output/genotypes.likelihoods.tsv",header=T,sep="\t") %>%
      group_by(sample,label) %>% 
      filter(r==max(r)) %>%
      filter(!grepl("trio",sample))
#exclude trios because they are related

sample_counts = read.table("./output/genotypes.likelihoods.tsv",header=T,sep="\t") %>%
  separate(sample,c("dataset","ID","p1","p2"),sep="\\.",remove=FALSE) %>%
  dplyr::select(dataset,ID) %>%
  unique() %>%
  group_by(dataset) %>%
  summarize(n=n())

n_modern = sample_counts %>% 
              filter(!dataset %in% c("ExcoffierAncient", "StoneAgeAncient","Archaic")) %>% 
              filter(!dataset %in% c("GTEx")) %>% 
              ungroup() %>% 
  summarize(n=sum(n))

sample_counts
n_modern

p1_counts = read.table("./output/genotypes.likelihoods.tsv",header=T,sep="\t") %>%
  separate(sample,c("dataset","ID","p1","p2"),sep="\\.",remove=FALSE) %>%
  filter(!dataset %in% c("ExcoffierAncient", "StoneAgeAncient","Archaic")) %>% 
  filter(!dataset %in% c("GTEx")) %>% 
  dplyr::select(dataset,ID,p1,p2) %>%
  group_by(p1) %>%
  summarize(n=n())
  
p1_counts %>% arrange()



t_gt_count = read.table("./output/genotypes.likelihoods.tsv",header=T,sep="\t") %>%
      group_by(sample,label) %>% 
      filter(r==max(r)) %>%
    filter(label=="AMY1") %>%
    separate(sample,c("dataset","ID",'p1','p2'),sep='\\.') 

t_gt  %>% filter(grepl("Excof",sample))
```




```{r}
## plot distributions of copy numbers for Figure 1

t_plot = t_gt %>% 
          filter(label%in% c("AMY1","AMY2A","AMY2B")) %>% 
          separate(sample,c("dataset","ID",'p1','p2'),sep='\\.') %>%
          filter(!dataset %in% c("StoneAgeAncient", "ExcoffierAncient","GTEx")) %>%
          group_by(p2,label) %>%
          mutate(p2_mu = mean(cp)) %>%
          ungroup() %>%
          group_by(p1,p2,p2_mu,label) %>%
          mutate(n_total=n(),
                 mu = mean(cp),
                 sdev = sd(cp)) %>%
          mutate(serr = sdev/sqrt(n_total)) %>%
          mutate(plabel = paste(p1," (",n_total,")",sep="")) %>%
          group_by(p1,plabel,cp,p2,label,n_total,mu,serr,p2_mu) %>%
          summarize(n=n()) %>%
          mutate(prop=n/n_total) %>%
          mutate(super_pop = ifelse(p1 %in% c("Denisova","Neanderthal"), "archaic","modern")) %>%
          mutate(p2 = ifelse(super_pop=="archaic","ARC",p2)) 


#c("ARC", "AFR", "OCN", "WEA", "SA", "CAS", "EA", "AMR")
t_plot$p2 = factor(t_plot$p2, levels=c("ARC", "AFR", "WEA", "SA", "CAS", "EA", "OCN", "AMR"))
#pnw_palette("Bay",6)
colors = c("black", "#EDB829", "#94B669", "#E97C07", "#00496F", "#0C7996","purple", "#DD4124")

pop_list = t_gt %>% 
          filter(label%in% c("AMY1","AMY2A","AMY2B")) %>% 
          separate(sample,c("dataset","ID",'p1','p2'),sep='\\.') %>%
          filter(dataset!="StoneAgeAncient") %>%
          ungroup() %>%
          dplyr::select(p1,p2) %>%
          unique()

outdir="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/read_depth_genotyping/genotyping/metadata"
fn_out = paste(outdir,"/populations.tsv",sep="")
write.table(pop_list,fn_out,sep="\t",row.names = FALSE,quote=FALSE)


t_plot %>% 
  ungroup() %>%
  dplyr::select(p2,label,p2_mu) %>% 
  unique() %>% 
  filter(label=="AMY1") %>% 
  arrange(-p2_mu)

```
```{r}
library(RColorBrewer)
#install.packages("ggsignif")
library(ggsignif)

t_subsistence = read.table("metadata/populations_subsistence.tsv",header=T,sep="\t") %>% 
                dplyr::select(type,p1)


exist = c("Yakut","Aleut","Tlingit","Mansi")
t_subsistence %>% filter(p1 %in% exist)
all_CAS = unique(t_gt %>% ungroup() %>% separate(sample,c("dataset","ID",'p1','p2'),sep='\\.') %>% filter(p2=="CAS") %>% dplyr::select(p1))$p1

#Biaka, Mbuti come from Perry

additional_info_df = data.frame(p1=c("Tubular",
                                    "Altaian",
                                    "Ulchi",
                                    "Chukchi",
                                    "EskimoSireniki",
                                    "EskimoChaplin",
                                    "EskimoNaukan",
                                    "Even",
                                    "Itelman",
                                    "Mansi",
                                    "Biaka",
                                    "Mbuti"),
                               type=c("Hunting",
                                      "Pastoralism",
                                      "Fishing",
                                      "Hunting",
                                      "Hunting",
                                      "Hunting",
                                      "Hunting",
                                      "Hunting",
                                      "Fishing",
                                      "Fishing",
                                      "Hunting",
                                      "Hunting"))

# "Tubular" Hunting https://en.wikipedia.org/wiki/Tubalars
# "Altaian" Pastoralism https://en.wikipedia.org/wiki/Altai_people
#"Ulchi" Fishing https://en.wikipedia.org/wiki/Ulch_people 
#"Chukchi" Hunting https://en.wikipedia.org/wiki/Chukchi_people
#"EskimoSireniki" Hunting https://scholarworks.alaska.edu/handle/11122/8302 https://www.erudit.org/en/journals/etudinuit/2007-v31-n1-2-etudinuit2570/019717ar/
#"EskimoChaplin" Hunting https://www.erudit.org/en/journals/etudinuit/2007-v31-n1-2-etudinuit2570/019717ar/
#"EskimoNaukan" Hunting https://en.wikipedia.org/wiki/Naukan_people
#"Even" Hunting https://www.jstor.org/stable/3773857?seq=4
#"Itelman"Fishing #https://www.jstor.org/stable/3773857?seq=4
#"Mansi" Fishing https://www.jstor.org/stable/3773857?seq=4
#"Kyrgyz"        
#"Mongolian" N/A

t_subsistence = rbind(t_subsistence,additional_info_df)

t_data_by_sub = t_gt %>% 
          filter(label%in% c("AMY1","AMY2A","AMY2B")) %>% 
          separate(sample,c("dataset","ID",'p1','p2'),sep='\\.') %>%
          inner_join(t_subsistence, on="p1") %>%
          #left_join(t_subsistence, on="p1") %>%
          #mutate(type = ifelse(is.na(type),"other",type)) %>%
          mutate(type=case_when(type=="Fishing" ~ "Fishing/Hunting",
                                type=="Hunting" ~ "Fishing/Hunting",
                                TRUE ~ type)) %>%
          mutate(simple_type = ifelse(type=="Agriculture",type,"other")) %>%
          filter(type!="Two or more sources")
          
t_plot_by_sub = t_data_by_sub %>%
                ungroup() %>%
                group_by(type,label,simple_type) %>%
                mutate(n_total=n(),
                    mu = mean(cp),
                    sdev = sd(cp)) %>%
          mutate(serr = sdev/sqrt(n_total)) %>%
          mutate(type_label = paste(type," (",n_total,")",sep="")) %>%
          group_by(type_label,cp,label,n_total,mu,serr,type,simple_type) %>%
          summarize(n=n()) %>%
          mutate(prop=n/n_total)

# t_sum_stats = t_plot_by_sub %>% 
#               ungroup() %>%
#               dplyr::select(type_label,mu,serr,type,label,simple_type) %>%
#               unique()

t_sum_stats = t_data_by_sub %>%
                ungroup() %>%
                group_by(simple_type,label) %>%
                summarize(n_total=n(),
                    mu = mean(cp),
                    sdev = sd(cp)) %>%
          mutate(serr = sdev/sqrt(n_total)) %>%
          mutate(x=ifelse(simple_type=="Agriculture",3,1.5))

p_A1 = wilcox.test(cp~simple_type, t_data_by_sub %>% filter(label=="AMY1"))$p.value
p_A2A = wilcox.test(cp~simple_type, t_data_by_sub %>% filter(label=="AMY2A"))$p.value
p_A2B = wilcox.test(cp~simple_type, t_data_by_sub %>% filter(label=="AMY2B"))$p.value
# p_A1 = t.test(cp~simple_type, t_data_by_sub %>% filter(label=="AMY1"))$p.value
# p_A2A = t.test(cp~simple_type, t_data_by_sub %>% filter(label=="AMY2A"))$p.value
# p_A2B = t.test(cp~simple_type, t_data_by_sub %>% filter(label=="AMY2B"))$p.value
# p_A1 = t.test(cp~simple_type, t_data_by_sub %>% filter(label=="AMY1",type!="Pastoralism"))$p.value
# p_A2A = t.test(cp~simple_type, t_data_by_sub %>% filter(label=="AMY2A",type!="Pastoralism"))$p.value
# p_A2B = t.test(cp~simple_type, t_data_by_sub %>% filter(label=="AMY2B",type!="Pastoralism"))$p.value

# mutate(p_adj = p.adjust(p.value)) %>%
#               mutate(p2 = str_replace(contrast,"WEA_","")) %>%
#               mutate(p2 = str_replace(p2,"_WEA","")) %>%
#               mutate(label = ifelse(p_adj>0.05,"N.S",
#                                     paste("P[adjusted]==",signif(p_adj,2)))) %>%

df_labs = data.frame(label = c("AMY1","AMY2A","AMY2B"),
                     ps = c(p_A1, p_A2A,p_A2B),
                     #xs = c(16.1,4.1,4.05),
                     xs = c(16.7,4.3,4.15),
                     #xs = c(16,4.5,3.75),
                     x_nudge = c(19,5,4.5)) %>%
            mutate(P_adj = p.adjust(ps)) %>%
            mutate(plabel=paste("P[adj]==",signif(P_adj,2)))

t_plot_by_sub$type = factor(t_plot_by_sub$type,levels=rev(c("Agriculture","Pastoralism","Fishing/Hunting","Two or more sources")))
t_plot_by_sub$type_label = factor(t_plot_by_sub$type_label,levels=rev(unique(t_plot_by_sub$type_label)))


# my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
# ggboxplot(ToothGrowth, x = "dose", y = "len",
#           color = "dose", palette = "jco")+ 
#   stat_compare_means(comparisons = my_comparisons)+ # Add pairwise comparisons p-value
#   stat_compare_means(label.y = 50)     # Add global p-value

#my_comparisons <- list( c("2", "1"), c("1", "2"), c("0.5", "2") )

mult_1=1.05
mult_2=1.07
df_segs = data.frame(x_start = rep(c(mult_1, mult_2, mult_1, mult_1),3),
                     x_end = rep(c(mult_1, mult_2, mult_2, mult_2),3),
                     y_start = rep(c(1, 1.5, 1.5,3),3),
                     y_end = rep(c(2, 3, 1.5,3),3),
                     label=rep(c("AMY1","AMY2A","AMY2B"),each=4)) %>%
  inner_join(data.frame(label=c("AMY1","AMY2A","AMY2B"),mult=c(15.8,4,3.9)),by="label")

g=ggplot(t_plot_by_sub)
g=g+geom_point(aes(x=type_label,y=cp,size=prop,color=type),alpha=0.8)+
  #geom_errorbar(aes(x=type_label,ymin=mu-serr,ymax=mu+serr),size=1,width=0,position=position_nudge(x=0.3),data=t_sum_stats)+
  #geom_point(aes(x=type_label,y=mu),shape=18,size=2,position=position_nudge(x=0.3),data=t_sum_stats)+
  #geom_point(aes(x=type_label,y=mu),color="white",shape=18,size=1,position=position_nudge(x=0.3),data=t_sum_stats)+
  geom_errorbar(aes(x=x,ymin=mu-serr,ymax=mu+serr),size=1,width=0,data=t_sum_stats)+
  geom_point(aes(x=x,y=mu),shape=18,size=2,data=t_sum_stats)+
  geom_point(aes(x=x,y=mu),color="white",shape=18,size=1,data=t_sum_stats)+
  geom_text(aes(x=1.0,y=xs,label=plabel),data=df_labs,parse=TRUE,size=1.75,hjust=0)+
  theme_bw(base_size=10)+
  facet_wrap(~label,ncol=1,strip.position="right",scales="free_x")+
  scale_color_manual(values=brewer.pal(7,"Spectral")[c(1,2,7)])+
  theme(legend.position="None")+
  geom_blank(aes(y=x_nudge,x=1),data=df_labs)+
  # geom_signif(
  #   y_position = c(5.3), 
  #   xmin = c(0), 
  #   xmax = c(1)
  # ) +
  geom_segment(aes(y=x_start*mult,yend=x_end*mult,x=y_start,xend=y_end),data=df_segs,lineend = "round",size=.25)+
  coord_flip()+
  scale_x_discrete("")+
  scale_y_continuous("",breaks=c(c(0,1,2,3,4,5),seq(6,28,2)))+
  theme(strip.background = element_blank())+
  scale_size(range=c(2,5))
  

fn_out = "./output_figures/main_figs/copy_by_sustinence.pdf"
pdf(fn_out,width=4,height=2.25)
print(g)
dev.off()


fn_out = "./output_figures/main_figs/copy_by_sustinence_wide.pdf"
pdf(fn_out,width=8,height=1.25)
print(g+facet_wrap(~label,nrow=1,strip.position="bottom",scales="free_x")+theme(strip.placement="outside"))
dev.off()

sus_pop_inf = t_data_by_sub %>% 
  ungroup() %>% 
  dplyr::select(p1,type,ID) %>% 
  unique() %>%
  group_by(p1,type) %>%
  summarize(n=n()) %>%
  arrange(type) 

sus_pop_inf %>% 
  ungroup() %>% 
  group_by(type) %>% 
  summarize(n=sum(n)) %>%
  mutate(n_total=sum(n))



```


```{r}
t_plot$super_pop = factor(t_plot$super_pop, levels = c("modern","archaic"))

t_plot_subset = t_plot %>% 
                  filter((n_total>=0)|(super_pop=="archaic"))
                  
group_orders = t_plot_subset %>% 
                ungroup %>% 
                dplyr::select(label,plabel,mu,p2,label) %>% 
                unique() %>%
                group_by(label,p2) %>%
                arrange(mu) %>%
                mutate(within_p2_order = row_number()) %>%
                ungroup() %>%
                dplyr::select(label,plabel,within_p2_order)

t_plot_subset = inner_join(t_plot_subset,group_orders,by=c("label","plabel")) %>%
                mutate(plabel=ifelse(within_p2_order%%2==1,paste(plabel,"-   -   -   -   -   -   -"),plabel)) 
min_size = 0.01
max_size = 1
b_size = 4

plot_amy = function(t_data, curr_label, label_x){
  t_sum = t_data %>% 
            dplyr::select(p1,plabel,p2,label,mu,serr,super_pop) %>%
            unique() %>%
            filter(label==curr_label)
  
  df_vline = data.frame(x=c(min((t_sum %>% filter(super_pop!="archaic"))$mu),
                            max((t_sum %>% filter(super_pop!="archaic"))$mu)),
                        super_pop = c('modern','modern'))
  
  df_vline$super_pop = factor(df_vline$super_pop, levels = c("modern","archaic"))
  df_vline_p2mu = t_plot_subset %>% filter(label==curr_label) %>% dplyr::select(p2,p2_mu,label) %>% unique()
  facet_labels = df_vline_p2mu %>% dplyr::select(p2) %>% mutate(x=label_x,y=1.5) %>% unique()
  
  order = (t_plot_subset %>% ungroup() %>% filter(label==curr_label) %>% dplyr::select(p2,p2_mu) %>% arrange(-p2_mu) %>% unique())$p2
  t_plot_subset$p2 = factor(t_plot_subset$p2,levels = order)
  
  colors = c("black", "#EDB829", "#94B669", "#E97C07", "#00496F", "#0C7996","purple", "#DD4124")
  colors = colors[match(order,c("ARC", "AFR", "WEA", "SA", "CAS", "EA", "OCN", "AMR"))]
  
  g=ggplot(t_plot_subset %>% filter(label == curr_label))
  g=g+geom_point(aes(x=cp,y=reorder(plabel,mu),size=prop,color=p2),alpha=.5)+
    #geom_errorbarh(aes(xmin=mu-serr,xmax=mu+serr,y=reorder(plabel,mu)),size=.05,height=0,data=t_sum)+
    geom_point(aes(x=mu,y=reorder(plabel,mu)),shape=18,data=t_sum,size=.5)+
    geom_point(aes(x=mu,y=reorder(plabel,mu),color=p2),shape=18,data=t_sum,size=.05,color='white')+
    geom_vline(aes(xintercept=x),data=df_vline,linetype='85',size=.1,alpha=.75)+
    geom_vline(aes(xintercept=p2_mu),data=df_vline_p2mu,linetype='63',size=.1,alpha=.75,color='red')+
    theme_bw(base_size=b_size)+
    theme(panel.grid = element_blank())+
    scale_color_manual("",values=colors)+
    theme(legend.key.size=unit(0.5,"cm"))+
    scale_size(range=c(min_size,max_size))+
    scale_x_continuous("",breaks=seq(0,20,1))+
    facet_grid(p2~.,space="free",scales="free_y")+
    theme(strip.background = element_blank(),
          strip.text = element_text(size=0),
          legend.position="None",
          panel.spacing=unit(0.02,"cm"))+
    geom_text(aes(x=x,label=p2,y=y),data=facet_labels,size=1)+
    scale_y_discrete("")
  
}

g = plot_amy(t_plot_subset, "AMY1", 18)
fn_out = "./output_figures/main_figs/genotype_by_pop_1.pdf"
pdf(fn_out,width=2,height=4)
print(g)
dev.off()

g = plot_amy(t_plot_subset, "AMY2A", 5)
fn_out = "./output_figures/main_figs/genotype_by_pop_2A.pdf"
pdf(fn_out,width=1.5,height=4)
print(g)
dev.off()

g = plot_amy(t_plot_subset, "AMY2B", 6)
fn_out = "./output_figures/main_figs/genotype_by_pop_2B.pdf"
pdf(fn_out,width=1.5,height=4)
print(g)
dev.off()

fn_out = "./output_figures/main_figs/genotype_by_pop_legend.pdf"
pdf(fn_out,width=1,height=4)
print(g+theme(legend.position="right",legend.key.size=unit(0.25,"cm")))
dev.off()

```
```{r}
##### PLOT BY SUPERPOP

t_data_by_superpop = t_gt %>% 
          filter(label%in% c("AMY1","AMY2A","AMY2B")) %>% 
          separate(sample,c("dataset","ID",'p1','p2'),sep='\\.') %>%
          filter(!dataset %in% c("StoneAgeAncient","ExcoffierAncient","GTEx")) %>%
          mutate(super_pop = ifelse(p1 %in% c("Denisova","Neanderthal"), "archaic","modern")) %>%
          mutate(p2 = ifelse(super_pop=="archaic","ARC",p2)) 
          
t_plot_by_superpop = t_data_by_superpop %>%
                ungroup() %>%
                group_by(p2,label) %>%
                mutate(n_total=n(),
                    mu = mean(cp),
                    sdev = sd(cp)) %>%
          mutate(serr = sdev/sqrt(n_total)) %>%
          mutate(p2_label = paste(p2," (",n_total,")",sep="")) %>%
          group_by(p2_label,cp,label,n_total,mu,serr,p2) %>%
          summarize(n=n()) %>%
          mutate(prop=n/n_total)
###

plot_amy_super = function(data, curr_label){
  t_data = data %>% filter(label==curr_label)
  
  t_sum = t_data %>% 
                ungroup() %>%
                dplyr::select(p2_label,mu,serr,label,p2) %>%
                unique()
  
  t_data$p2 = factor(t_data$p2, levels=c("ARC", "AFR", "WEA", "SA", "CAS", "EA", "OCN", "AMR"))
  
  order = (t_data %>% ungroup() %>% filter(label==curr_label) %>% dplyr::select(p2,mu) %>% arrange(-mu) %>% unique())$p2
  colors = c("black", "#EDB829", "#94B669", "#E97C07", "#00496F", "#0C7996","purple", "#DD4124")
  
  g=ggplot(t_data)
  g=g+geom_point(aes(x=cp,y=reorder(p2_label,mu),size=prop,color=p2),alpha=.5)+
     #facet_wrap(~label,shrink = TRUE)+
      geom_point(aes(x=mu,y=reorder(p2_label,mu)),shape=18,data=t_sum,size=1.5,position=position_nudge(x=0,y=0))+
      geom_point(aes(x=mu,y=reorder(p2_label,mu),color=p2),shape=18,data=t_sum,size=.5,position=position_nudge(x=0,y=0),color='white')+
      theme_bw(base_size=8)+
      scale_color_manual("",values=colors)+
      theme(legend.key.size=unit(0.5,"cm"))+
      scale_size(range=c(0.01,3.5))+
      scale_x_continuous("",breaks=seq(0,20,2))+
      theme(strip.background = element_blank(),
            strip.text = element_text(size=3),
            legend.position="None",
            panel.grid=element_blank())+
      scale_y_discrete("")
  
   return(g)
}
#    geom_text(aes(x=x,y=plabel, label=label),parse=TRUE,data=p_values,size=1,position=position_nudge(x=0,y=.35),hjust=1)

g = plot_amy_super(t_plot_by_superpop, "AMY1")
  
fn_out = "./output_figures/main_figs/genotype_by_superpop_AMY1.pdf"
pdf(fn_out,width=2,height=1.5)
print(g)
dev.off()

g = plot_amy_super(t_plot_by_superpop, "AMY2A")
  
fn_out = "./output_figures/main_figs/genotype_by_superpop_AMY2A.pdf"
pdf(fn_out,width=1.5,height=1.5)
print(g)
dev.off()

g = plot_amy_super(t_plot_by_superpop, "AMY2B")
fn_out = "./output_figures/main_figs/genotype_by_superpop_AMY2B.pdf"
pdf(fn_out,width=1.5,height=1.5)
print(g)
dev.off()

```
```{r}
library(ggforce)
#### plot jit AMYX vs AMYY

t_struc = t_data_by_superpop %>% 
  #dplyr::select(dataset,ID,p1,p2,X,cp,label) %>%
  dplyr::select(dataset,ID,p1,p2,cp,label) %>%
  pivot_wider(values_from = cp, names_from = label)

t_struc$p2 = factor(t_struc$p2, levels=c("ARC", "AFR", "WEA", "SA", "CAS", "EA", "OCN", "AMR"))
  
colors = c("black", "#EDB829", "#94B669", "#E97C07", "#00496F", "#0C7996","purple", "#DD4124")

# g=ggplot(t_struc)
# g=g+geom_point(aes(x=AMY1,y=AMY2B), 
#                position=position_jitternormal(sd_x = 0.1, sd_y=0.1),
#                alpha=.8,
#                size=0.01,
#                shape=16)+
#   theme_bw(base_size=8)+
#   scale_color_manual(values=colors)+
#   theme(legend.position="None",
#         panel.grid=element_blank())+
#   scale_x_continuous("AMY1",breaks=seq(0,20,2))
# 
# fn_out = "./output_figures/main_figs/copy_compare_AMY1_vs_AMY2B.pdf"
# pdf(fn_out,width=2,height=1)
# print(g)
# dev.off()
# 
# g=ggplot(t_struc %>% arrange(p2))
# g=g+geom_point(aes(x=AMY2A,y=AMY2B,), 
#                position=position_jitternormal(sd_x = 0.1, sd_y=0.1),
#                alpha=.8,
#                shape=16,
#                size=0.01)+
#   theme_bw(base_siz e=8)+
#   scale_color_manual(values=colors)+
#   theme(legend.position="None",
#         panel.grid=element_blank())+
#   scale_x_continuous("AMY2A",breaks=seq(0,6))
# 
# fn_out = "./output_figures/main_figs/copy_compare_AMY2A_vs_AMY2B.pdf"
# pdf(fn_out,width=1.25,height=1)
# print(g)
# dev.off()


library(scales)
norm_trans <- function(){
  trans_new('norm', function(x) log10(x), function(x) 10^(x))
}

#badVersion + geom_point() + scale_colour_continuous(trans = 'norm'))

t_1_v_2B_sum = t_struc %>%
  group_by(AMY1,AMY2B) %>%
  summarize(n=n())

g=ggplot(t_1_v_2B_sum)
g=g+geom_point(aes(x=AMY1,y=AMY2B,size=n,color=n))+
  theme_bw(base_size=8)+
  scale_color_manual(values=colors)+
  theme(legend.position=c(0.9,0.8),
        legend.key.size=unit(0.05,"cm"),
        legend.text = element_text(size=3),
        legend.title = element_text(size=4),
        legend.background = element_blank(),
        panel.grid=element_blank())+
  scale_x_continuous("AMY1",breaks=seq(0,20,2))+
#  scale_color_viridis_c("", option="inferno",trans = 'norm')+
#  scale_size_continuous(range=c(0.01,3))+
  scale_color_viridis_c("", option="inferno",trans = 'norm',breaks=c(100,250,500,750))+
  scale_size_continuous(range=c(0.01,3),breaks=c(100,250,500,750))+
  geom_blank(aes(x=4,y=1.5))+
  guides(color = guide_legend(""),
         size = guide_legend(""))

fn_out = "./output_figures/main_figs/copy_compare_AMY1_vs_AMY2B_heatmap.pdf"
pdf(fn_out,width=2,height=1)
print(g)
dev.off()

t_2A_v_2B_sum = t_struc %>%
  group_by(AMY2A,AMY2B) %>%
  summarize(n=n())

g=ggplot(t_2A_v_2B_sum)
g=g+geom_point(aes(x=AMY2A,y=AMY2B,color=n,size=n))+
  theme_bw(base_size=8)+
  scale_color_manual(values=colors)+
  theme(legend.position=c(0.15,0.8),
        legend.key.size=unit(0.05,"cm"),
        legend.text = element_text(size=3),
        legend.title = element_text(size=4),
        legend.background = element_blank(),
        panel.grid=element_blank())+
  scale_x_continuous("AMY2A",breaks=seq(0,20,1))+
  geom_blank(aes(x=0,y=1.5))+
  scale_color_viridis_c("", option="inferno",trans = 'norm',breaks=c(100,500,1000,2000))+
  scale_size_continuous(range=c(0.01,3),breaks=c(100,500,1000,2000))+
  guides(color = guide_legend(""),
         size = guide_legend(""))

fn_out = "./output_figures/main_figs/copy_compare_AMY2A_vs_AMY2B_heatmap.pdf"
pdf(fn_out,width=1.5,height=1)
print(g)
dev.off()



```


```{r}

############## TO ADAPT

#for (curr_gene in c("AMY1","AMY2B","AMY2A")){


t_gt_renamed = t_gt %>% 
          filter(label%in% c("AMY1","AMY2A","AMY2B")) %>% 
          separate(sample,c("dataset","ID",'p1','p2'),sep='\\.') %>%
          mutate(p2 = ifelse(p2=="HunterGathererE","Hunter-GathererE",p2)) %>%
          filter(p2 %in% c("WEA","EarlyFarmer","Hunter-GathererW","Hunter-GathererE","Yamnaya","Europe")) %>%
          mutate(super_pop = ifelse(p1 %in% c("Denisova","Neanderthal"), "archaic","modern")) %>%
          mutate(super_pop = ifelse(p2 %in% c("EarlyFarmer","Hunter-GathererW","Hunter-GathererE","Yamnaya"), "ancient",super_pop)) %>%
          mutate(p2 = ifelse(super_pop=="archaic","Archaic",p2)) %>%
          group_by(p2,label,super_pop) %>%
          mutate(plabel = paste(p2," (",n(),")",sep="")) %>%
          ungroup()
          

t_plot =  t_gt_renamed %>%
          group_by(p2,label,super_pop,plabel) %>%
          mutate(n_total=n(),
                 mu = mean(cp),
                 sdev = sd(cp)) %>%
          mutate(serr = sdev/sqrt(n_total)) %>%
          group_by(p2,plabel,cp,label,n_total,mu,serr,super_pop) %>%
          summarize(n=n()) %>%
          mutate(prop=n/n_total) %>%
          mutate(order = case_when(p2 == "WEA" ~ 1,
                                  p2 == "Yamnaya" ~ 2,
                                  p2 == "EarlyFarmer" ~ 3,
                                  p2 == "Hunter-GathererW" ~ 4,
                                  p2 == "Hunter-GathererE" ~ 5,
                                  p2 == "Archaic" ~ 6)) %>%
          mutate(order = -1*order)

t_labels = t_plot %>% ungroup() %>% dplyr::select(p2,plabel) %>% unique()

curr_gene = "AMY1"



#"Archaic"          "EarlyFarmer"      "Hunter-GathererE" "Hunter-GathererW" "WEA"              "Yamnaya"  
t_plot$p2 = factor(t_plot$p2, levels=c("Archaic", "WEA", "Hunter-GathererE", "Hunter-GathererW", "EarlyFarmer","Yamnaya"))
colors = c("black",  "#94B669", pnw_palette(name="Starfish",n=4,type="discrete"))

WEAs = c("Archaic",
         "WEA",
         "Hunter-GathererE",
         "Hunter-GathererW",
         "Yamnaya",
         "EarlyFarmer")

combos = unique(WEAs) %>%
  combn(2, simplify = F) %>%
  set_names(map_chr(., ~ paste(., collapse = "_"))) 


lims = data.frame(gene=c("AMY1","AMY2A","AMY2B"),
                  mn = c(2,0,0),
                  mx = c(20,7,7))

for (curr_gene in c("AMY1","AMY2A","AMY2B")){
  t_sum_plot = t_plot %>% filter(label==curr_gene)
  t_sum_all_data = t_gt_renamed %>% filter(label==curr_gene)
  
  fn_out = paste("./output_figures/ancient_trajectories/genotype_by_pop_ancient_",curr_gene,".pdf",sep="")
  

  p_values = map_df(combos, function(y) {
                filter(t_sum_all_data, p2 %in% y) %>% 
                  wilcox.test(cp ~ p2, data = .) %>% 
                  broom::tidy()
                }, .id = "contrast")  %>%
              filter(grepl("WEA",contrast)) %>%
              mutate(p_adj = p.adjust(p.value)) %>%
              mutate(p2 = str_replace(contrast,"WEA_","")) %>%
              mutate(p2 = str_replace(p2,"_WEA","")) %>%
              mutate(label = ifelse(p_adj>0.05,"N.S",
                                    paste("P[adjusted]==",signif(p_adj,2)))) %>%
              mutate(p3="European") %>%
              mutate(x=case_when(
                curr_gene=="AMY1" ~ 20,
                curr_gene=="AMY2B" ~ 7,
                curr_gene=="AMY2A" ~ 7,
                TRUE ~0
              )) %>%
      inner_join(t_labels,by="p2")
  
  
 t_sum = t_plot %>% 
          dplyr::select(p2,plabel,label,mu,serr,super_pop,order) %>%
          unique() %>%
          filter(label==curr_gene)
          
  mn = (lims %>% filter(gene==curr_gene))$mn
  mx = (lims %>% filter(gene==curr_gene))$mx
  
  g=ggplot(t_plot %>% filter(label == curr_gene))
  g=g+geom_point(aes(x=cp,y=reorder(plabel,order),size=prop,color=p2),alpha=.5)+
    geom_errorbarh(aes(xmin=mu-serr,xmax=mu+serr,y=reorder(plabel,order)),size=.05,height=0,data=t_sum,position=position_nudge(x=0,y=0))+
    geom_point(aes(x=mu,y=reorder(plabel,order)),shape=18,data=t_sum,size=1.5,position=position_nudge(x=0,y=0))+
    geom_point(aes(x=mu,y=reorder(plabel,order),color=p2),shape=18,data=t_sum,size=.5,position=position_nudge(x=0,y=0))+
    #stat_summary(aes(x=gt,y=reorder(p1,gt),color=p2),size=.1)+
    theme_bw(base_size=4)+
    scale_color_manual("",values=colors)+
    theme(legend.key.size=unit(0.5,"cm"))+
    scale_size(range=c(0.01,3.5))+
    scale_x_continuous(paste(curr_gene," copy",sep=" "),breaks=seq(0,20,2),lim=c(mn,mx))+
    #facet_grid(super_pop~.,space="free",scales="free_y")+
    theme(strip.background = element_blank(),
          strip.text = element_text(size=1),
          legend.position="None")+
    scale_y_discrete("")+
    geom_text(aes(x=x,y=plabel, label=label),parse=TRUE,data=p_values,size=1,position=position_nudge(x=0,y=.35),hjust=1)
  
  
  #fn_out = "./output_figures/genotype_by_pop_ancient_2B.pdf"
  pdf(fn_out,width=2,height=0.75)
  print(g)
  dev.off()
  

}
```
```{r}

library(ggpmisc)
#install.packages("ggpmisc")

ancient_info = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/haplotype_analysis/metadata/ancient_info.tsv",header=T,sep="\t") %>%
  mutate(ID = sample, age=ageAverage) %>%
  dplyr::select(ID, age)

ancient_info_excof = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/read_depth_genotyping/d4_cvg_analysis/bam_to_d4/meta_data/ExcoffierAncients.tsv",header=T,sep="\t") %>%
  mutate(ID = SAMPLE_ID, age=AGE_AVG_BP) %>%
  dplyr::select(ID, age)

ancient_info = rbind(ancient_info, ancient_info_excof)

lmp <- function (modelobject) {
    if (class(modelobject) != "lm") stop("Not an object of class 'lm' ")
    f <- summary(modelobject)$fstatistic
    p <- pf(f[1],f[2],f[3],lower.tail=F)
    attributes(p) <- NULL
    return(p)
}


t_gt_renamed_age = left_join(t_gt_renamed,ancient_info,by="ID") %>%
               mutate(age=ifelse(!dataset %in% c("StoneAgeAncient","ExcoffierAncient"),0,as.numeric(age))) %>%
               #filter(sample!="StoneAgeAncient.VK117.NorthernEurope.Yamnaya") %>%
               mutate(p2 = ifelse(dataset=="Archaic","Archaic",p2)) %>%
               mutate(age=ifelse(super_pop=="modern",0,age)) %>%
               filter(age<20000) %>%
               filter(p2!="archaic")
               

t_gt_renamed_age$p2 = factor(t_gt_renamed_age$p2, levels=c("WEA", "Hunter-GathererE", "Hunter-GathererW", "EarlyFarmer","Yamnaya"))
colors = c("#94B669", pnw_palette(name="Starfish",n=4,type="discrete"))


curr_gene = "AMY1"
curr_gene = "AMY2A"
#curr_gene = "AMY2B"
inset_pos = data.frame(gene=c("AMY1","AMY2A","AMY2B"),y=c(18,8,8))
dummy = data.frame(gene=c("AMY1","AMY2A","AMY2B"),y=c(2,2,1.5))


for (curr_gene in c("AMY1","AMY2A","AMY2B")){

  fn_out = paste("./output_figures/ancient_trajectories/genotype_by_pop_ancient_corr",curr_gene,".pdf",sep="")
  
  breaks = seq(-15000,0,5000)
  labels = paste(breaks/1000)
  label = "kyr BP"
  

  g_i=ggplot(t_gt_renamed_age %>% filter(label == curr_gene))
  g_i=g_i+geom_smooth(aes(x=-age,y=cp),alpha=.5,color="blue",fill="lightblue",size=.5,method='lm')+
    geom_smooth(aes(x=-age,y=cp),alpha=.5,color="red",fill="pink",size=.5,method='gam')+
    scale_color_manual("",values=colors)+
    theme_bw(base_size=4)+
    theme(legend.position="None")+
    scale_y_continuous("")+
    scale_x_continuous("")+
    theme(axis.text = element_blank(), 
          panel.background = element_blank())+
    theme(axis.line = element_line(colour = "black"),
          #panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          #panel.border = element_blank(),
          panel.background = element_rect(fill='white'),
          plot.background=element_rect(fill='transparent', color=NA))
  
  inset_pos_y = (inset_pos %>% filter(gene==curr_gene))$y  
  dummy_y = data.frame(y=(dummy %>% filter(gene==curr_gene))$y,x=0)  
    
  mn = min((t_gt_renamed_age %>% filter(label==curr_gene))$cp)-.1
  mx = max((t_gt_renamed_age %>% filter(label==curr_gene))$cp)+.1
  
  data.tb <- tibble(x = -13500, y = mx+mx/25, 
                  plot = list(g_i))
  
  pval = lmp(lm(age~cp,t_gt_renamed_age %>% filter(label == curr_gene)))
  m = lm(cp~age,t_gt_renamed_age %>% filter(label == curr_gene) %>% mutate(age=-age))
  mu_pred = predict(m,data.frame(age=c(-12000,0)))
  print(curr_gene)
  delta = mu_pred[2]-mu_pred[1]
  print(c(mu_pred,delta))
  m2 = mgcv::gam(cp ~ s(age, bs = "cs"), data=t_gt_renamed_age %>% filter(label == curr_gene) %>% mutate(age=-age))
  summary(m2)
  mu_pred_m2 = predict(m2,data.frame(age=c(-12000,0)))
  delta = mu_pred_m2[2]-mu_pred_m2[1]
  print(c(mu_pred_m2,delta))
  
  df_p = data.frame(x=-6000,y=mx,label=paste("P==",signif(pval,2),sep=""))
  print(dummy_y)
  
  g=ggplot(t_gt_renamed_age %>% filter(label == curr_gene))
  g=g+geom_point(aes(x=-age,y=cp,color=p2),alpha=.5,position=position_jitter(width=0,height=.2),size=.5)+
    geom_plot(data = data.tb, aes(x, y, label = plot)) +
    geom_smooth(aes(x=-age,y=cp),alpha=.5,color="red",size=.5,method="gam")+
    scale_color_manual("",values=colors)+
    theme_bw(base_size=8)+
    theme(legend.position="None")+
    scale_y_continuous(paste(curr_gene," copy",sep=""),breaks=seq(0,20,2))+
    scale_x_continuous(label,breaks=breaks,labels=labels)+
    #coord_cartesian(xlim=c(-12000,0))+
    coord_cartesian(xlim=c(-12000,0),ylim=c(mn,mx))+
    geom_blank(aes(x=x,y=y),data=dummy_y)+
    geom_text(aes(x=x,y=y,label=label),data=df_p,parse=TRUE,size=2,hjust=0)

       
  pdf(fn_out,width=1.5,height=1.5)
  print(g)
  dev.off()
  print(g)
}

### NOTE THIS INDIVIDUAL
###**
left_join(t_gt_renamed,ancient_info,by="ID") %>%
               mutate(age=ifelse(dataset!="StoneAgeAncient",0,as.numeric(age))) %>%
               #filter(sample!="StoneAgeAncient.VK117.NorthernEurope.Yamnaya") %>%
               mutate(p2 = ifelse(dataset=="Archaic","Archaic",p2)) %>%
               mutate(age=ifelse(super_pop=="modern",0,age)) %>%
               filter(age>20000)


g=ggplot(t_gt_renamed_age)
g+geom_histogram(aes(x=age))

t_gt_renamed_age %>% 
  filter(age!=0) %>%
  dplyr::select(age,ID,p1,p2) %>%
  filter(age==min(age))
#11627.5
#250

```
```{r}



```



```{r}            
  g=ggplot(t_amy %>% filter(name==curr_gene))
  g=g+geom_density_ridges(aes(x=gt,y=p2,fill=p2,color=p2),
                        binwidth=.1,
                        alpha=.1,
                        size=.5,
                        scale=.7,adjust=.01)+
    geom_point(aes(x=gt,y=p2),shape="|",size=1,alpha=.25,position=position_nudge(y=-.25))+
    geom_errorbarh(aes(xmin=mu-sem,xmax=mu+sem,y=p2,color=p2),data=t_sum_plot,height=0.1)+
    geom_point(aes(x=mu,y=p2,color=p2),data=t_sum_plot,size=1)+
    geom_point(aes(x=mu,y=p2),data=t_sum_plot,size=.1,color="white")+
    geom_text(aes(x=x,y=p,label=label),
              parse=TRUE,
              data=p_values,
              size=2,
              hjust=0,
              position=position_nudge(y=.35))+
    theme_bw(base_size=6)+
    scale_x_continuous("copy genotype",breaks=seq(0,20,2))+
    scale_y_discrete("")+
    facet_grid(p3~.,shrink = TRUE,drop = TRUE,scales="free_y")+
    theme(strip.background=element_blank())+
    scale_color_manual(values=colors)+
    scale_fill_manual(values=colors)+
    theme(legend.position="None")
  pdf(fn_out,width=2.5,height=2)
  print(g)
  dev.off()  
  
}



```

```{r}
#plot RAW data and means


g=ggplot(t %>% filter(name%in% c("amy1_sum","AMY2A","AMY2B")) %>% filter(dataset!="StoneAgeAncient") %>% filter(name == "amy1_sum"))
g=g+
  #geom_point(aes(x=gt,reorder(p1,gt),color=p2),size=.5,alpha=.5,shape="|")+
  geom_point(aes(x=gt,reorder(p1,gt)),size=.5,alpha=.5,shape="|")+
  stat_summary(aes(x=gt,y=reorder(p1,gt),color=p2),size=.1)+
  theme_bw(base_size=4)+
  facet_wrap(~name,scales = "free")+
  scale_color_manual("",values=pnw_palette("Bay",8))+
  theme(legend.key.size=unit(0.5,"cm"))
  #scale_color_brewer(palette="BuGn")

fn_out = paste(outdir,"/copy_by_pop_1.pdf",sep="")
pdf(fn_out,width=2.5,height=4)
print(g)
dev.off()

g=ggplot(t %>% filter(name%in% c("amy1_sum","AMY2A","AMY2B")) %>% filter(dataset!="StoneAgeAncient") %>% filter(name == "AMY2A"))
g=g+
  #geom_point(aes(x=gt,reorder(p1,gt),color=p2),size=.5,alpha=.5,shape="|")+
  geom_point(aes(x=gt,reorder(p1,gt)),size=.5,alpha=.5,shape="|")+
  stat_summary(aes(x=gt,y=reorder(p1,gt),color=p2),size=.1)+
  theme_bw(base_size=4)+
  facet_wrap(~name,scales = "free")+
  scale_color_manual("",values=pnw_palette("Bay",8))+
  theme(legend.key.size=unit(0.5,"cm"))
  #scale_color_brewer(palette="BuGn")


fn_out = paste(outdir,"/copy_by_pop_2A.pdf",sep="")
pdf(fn_out,width=2.5,height=4)
print(g)
dev.off()


g=ggplot(t %>% filter(name%in% c("amy1_sum","AMY2A","AMY2B")) %>% filter(dataset!="StoneAgeAncient") %>% filter(name == "AMY2B"))
g=g+
  #geom_point(aes(x=gt,reorder(p1,gt),color=p2),size=.5,alpha=.5,shape="|")+
  geom_point(aes(x=gt,reorder(p1,gt)),size=.5,alpha=.5,shape="|")+
  stat_summary(aes(x=gt,y=reorder(p1,gt),color=p2),size=.1)+
  theme_bw(base_size=4)+
  facet_wrap(~name,scales = "free")+
  scale_color_manual("",values=pnw_palette("Bay",8))+
  theme(legend.key.size=unit(0.5,"cm"))
  #scale_color_brewer(palette="BuGn")


fn_out = paste(outdir,"/copy_by_pop_2B.pdf",sep="")
pdf(fn_out,width=2.5,height=4)
print(g)
dev.off()



```
```{r}

g=ggplot(t %>% filter(name%in% c("amy1_sum","AMY2A","AMY2B")) %>% filter(dataset!="StoneAgeAncient") %>% filter(name == "AMY2B"))
g=g+geom_point(aes(x=gt,reorder(p1,gt),color=p2),size=.5,alpha=.5,shape="|")+
  stat_summary(aes(x=gt,y=reorder(p1,gt),color=p2),size=.1)+
  theme_bw(base_size=4)+
  facet_wrap(~name,scales = "free")+
  scale_color_manual("",values=pnw_palette("Bay",8))+
  theme(legend.key.size=unit(0.5,"cm"))
  #scale_color_brewer(palette="BuGn")

fn_out = "/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/read_depth_genotyping/genotyping/output_figures/copy_by_pop_2B.pdf"
pdf(fn_out,width=2.5,height=4)
print(g)
dev.off()

```

```{r}
#t %>% filter(name%in% c("amy1_sum","AMY2A","AMY2B")) 
#  %>% filter(dataset!="StoneAgeAncient")
#  %>% spread(name, gt)

t %>% spread(name, gt)


```

```{r}
g=ggplot(t %>% spread(name, gt) %>% filter(dataset!="StoneAgeAncient"))
g=g+geom_point(aes(x=amy1_sum,y=AMY2A,color=p2),size=0.9)+
  #stat_summary(aes(x=gt,y=reorder(p1,gt),color=p2),size=.1)+
  #theme_bw(base_size=4)+
  #facet_wrap(~name,scales = "free")+
  scale_color_manual("",values=pnw_palette("Bay",8))+
  theme(legend.key.size=unit(0.5,"cm"))+
  ylim(-0.25,7.25)+
  xlim(0,21)+
  #scale_color_brewer(palette="BuGn")
 labs(title = "AMY2A vs. AMY1 Copy Number (1KG and HGDP)", x = "AMY1 Summed Copy Number", y = "AMY2A Copy Number")

fn_out = "./output_figures/AMY2A_vs_AMY1_scatterplot.pdf"
pdf(fn_out,width=8,height=6)
print(g)
dev.off()




```
```{r}

##########
###OLD###


# p1 = subpops
# p2 = superpops unique(t_plot$p1)
# p2 = early farmer etc

unique(t_gt$p2)

t_gt %>% filter(grepl("Nean",sample))

t_plot = t_gt %>% 
          filter(label%in% c("AMY1","AMY2A","AMY2B")) %>% 
          separate(sample,c("dataset","ID",'p1','p2'),sep='\\.') %>%
          filter(p2 %in% c("WEA","EarlyFarmer","Hunter-GathererW","Hunter-GathererE","Yamnaya","Europe")) %>%
          mutate(super_pop = ifelse(p1 %in% c("Denisova","Neanderthal"), "archaic","modern")) %>%
          mutate(super_pop = ifelse(p2 %in% c("EarlyFarmer","Hunter-GathererW","Hunter-GathererE","Yamnaya"), "ancient",super_pop)) %>%
          mutate(p2 = ifelse(super_pop=="archaic","Archaic",p2)) %>%
          group_by(p2,label,super_pop) %>%
          mutate(n_total=n(),
                 mu = mean(cp),
                 sdev = sd(cp)) %>%
          mutate(serr = sdev/sqrt(n_total)) %>%
          mutate(plabel = paste(p2," (",n_total,")",sep="")) %>%
          group_by(p2,plabel,cp,label,n_total,mu,serr,super_pop) %>%
          summarize(n=n()) %>%
          mutate(prop=n/n_total) %>%
          mutate(order = case_when(p2 == "WEA" ~ 1,
                                  p2 == "Yamnaya" ~ 2,
                                  p2 == "EarlyFarmer" ~ 3,
                                  p2 == "Hunter-GathererW" ~ 4,
                                  p2 == "Hunter-GathererE" ~ 5,
                                  p2 == "Archaic" ~ 6)) %>%
          mutate(order = -1*order)
          
# t_plot$p2 = factor(t_plot$p2,levels=c("WEA",
#                                       "Yamnaya",
#                                       "EarlyFarmer",
#                                       "Hunter-GathererW",
#                                       "Hunter-GathererE",
#                                       "Europe"))

t_sum = t_plot %>% 
          dplyr::select(p2,plabel,label,mu,serr,super_pop,order) %>%
          unique() %>%
          filter(label=="AMY1")
          

g=ggplot(t_plot %>% filter(label == "AMY1") )
g=g+geom_point(aes(x=cp,y=reorder(plabel,order),size=prop,color=p2),alpha=.5)+
  geom_errorbarh(aes(xmin=mu-serr,xmax=mu+serr,y=reorder(plabel,order)),size=.05,height=0,data=t_sum)+
  geom_point(aes(x=mu,y=reorder(plabel,order)),shape=18,data=t_sum,size=1.5)+
  geom_point(aes(x=mu,y=reorder(plabel,order),color=p2),shape=18,data=t_sum,size=.5)+
  #stat_summary(aes(x=gt,y=reorder(p1,gt),color=p2),size=.1)+
  theme_bw(base_size=4)+
  scale_color_manual("",values=colors)+
  theme(legend.key.size=unit(0.5,"cm"))+
  scale_size(range=c(0.01,1.5))+
  scale_x_continuous(breaks=seq(0,20,2))+
  #facet_grid(super_pop~.,space="free",scales="free_y")+
  theme(strip.background = element_blank(),
        strip.text = element_text(size=1),
        legend.position="None")+
  scale_y_discrete("")


fn_out = "./output_figures/genotype_by_pop_ancient_1.pdf"
pdf(fn_out,width=2,height=0.75)
print(g)
dev.off()


t_sum = t_plot %>% 
          dplyr::select(p2,plabel,label,mu,serr,super_pop,order) %>%
          unique() %>%
          filter(label=="AMY2A")
          

g=ggplot(t_plot %>% filter(label == "AMY2A") )
g=g+geom_point(aes(x=cp,y=reorder(plabel,order),size=prop,color=p2),alpha=.5)+
  geom_errorbarh(aes(xmin=mu-serr,xmax=mu+serr,y=reorder(plabel,order)),size=.05,height=0,data=t_sum)+
  geom_point(aes(x=mu,y=reorder(plabel,order)),shape=18,data=t_sum,size=1.5)+
  geom_point(aes(x=mu,y=reorder(plabel,order),color=p2),shape=18,data=t_sum,size=.5)+
  #stat_summary(aes(x=gt,y=reorder(p1,gt),color=p2),size=.1)+
  theme_bw(base_size=4)+
  scale_color_manual("",values=colors)+
  theme(legend.key.size=unit(0.5,"cm"))+
  scale_size(range=c(0.01,1.5))+
  scale_x_continuous(breaks=seq(0,20,2))+
  #facet_grid(super_pop~.,space="free",scales="free_y")+
  theme(strip.background = element_blank(),
        strip.text = element_text(size=1),
        legend.position="None")+
  scale_y_discrete("")


fn_out = "./output_figures/genotype_by_pop_ancient_2A.pdf"
pdf(fn_out,width=2,height=0.75)
print(g)
dev.off()




t_sum = t_plot %>% 
          dplyr::select(p2,plabel,label,mu,serr,super_pop,order) %>%
          unique() %>%
          filter(label=="AMY2B")
          

g=ggplot(t_plot %>% filter(label == "AMY2B") )
g=g+geom_point(aes(x=cp,y=reorder(plabel,order),size=prop,color=p2),alpha=.5)+
  geom_errorbarh(aes(xmin=mu-serr,xmax=mu+serr,y=reorder(plabel,order)),size=.05,height=0,data=t_sum)+
  geom_point(aes(x=mu,y=reorder(plabel,order)),shape=18,data=t_sum,size=1.5)+
  geom_point(aes(x=mu,y=reorder(plabel,order),color=p2),shape=18,data=t_sum,size=.5)+
  #stat_summary(aes(x=gt,y=reorder(p1,gt),color=p2),size=.1)+
  theme_bw(base_size=4)+
  scale_color_manual("",values=colors)+
  theme(legend.key.size=unit(0.5,"cm"))+
  scale_size(range=c(0.01,1.5))+
  scale_x_continuous(breaks=seq(0,20,2))+
  #facet_grid(super_pop~.,space="free",scales="free_y")+
  theme(strip.background = element_blank(),
        strip.text = element_text(size=1),
        legend.position="None")+
  scale_y_discrete("")


fn_out = "./output_figures/genotype_by_pop_ancient_2B.pdf"
pdf(fn_out,width=2,height=0.75)
print(g)
dev.off()





```














