---
title: "Untitled"
output: html_document
date: "2023-04-26"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
outdir="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/graph_genotyping/assess_graph_genotypes/output"

# t_clusts = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/HPRC_AMY_Sequences/combined_y1_y2_analyses/graph_genotyping_clusters/output/clusters_processed.tsv",header=TRUE,sep="\t",comment="") %>%
#   transmute(sample = str_replace(sample,"-","_"),cluster=cluster)

#t_clusts = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/HPRC_AMY_Sequences/combined_y1_y2_analyses/graph_genotyping_clusters/output/20230718_dendrogram.jaccard.bestcut_clusters_processed.tsv",header=TRUE,sep="\t",comment="") %>%
#  transmute(sample = str_replace(sample,"-","_"),cluster=cluster)

#t_clusts = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/HPRC_AMY_Sequences/combined_y1_y2_analyses/graph_genotyping_clusters/output/20230817_dendrogram.jaccard.bestcut_clusters_processed.tsv",header=TRUE,sep="\t",comment="") %>%
#  transmute(sample = str_replace(sample,"-","_"),cluster=cluster)

t_clusts = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/HPRC_AMY_Sequences/combined_y1_y2_analyses/output/haplotype_assignments.txt",header=TRUE,sep="\t",comment="") %>% 
  dplyr::select(-hap_struc_d) %>%
  mutate(chrom=str_replace(chrom,"-","_"))
  

# t = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/graph_genotyping/result_tsv/cosigt_results.bestgeno.tsv",sep="\t",comment.char = "",col.names = c("sample", "besthaplo1_besthaplo2",  "best_score", "cohort")) %>% separate(besthaplo1_besthaplo2,into=c("hap1","hap2"),sep="\\$",extra="merge") %>%
#   filter(cohort!="SGDP") %>%
#   mutate(cohort=ifelse(cohort=="SGDPv2","SGDP",cohort))

t = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/graph_genotyping/result_tsv/new_results/best_geno/best_geno_tot.tsv",sep="\t",comment.char = "",col.names = c("sample", "besthaplo1_besthaplo2",  "best_score", "cohort")) %>%
  separate(besthaplo1_besthaplo2,into=c("hap1","hap2"),sep="\\$",extra="merge") %>%
  filter(cohort!="SGDP") %>%
  mutate(cohort=ifelse(cohort=="SGDPv2","SGDP",cohort)) %>%
  mutate(cohort=ifelse(cohort=="1K","1KG",cohort)) %>%
  mutate(cohort=ifelse(cohort=="1KGrel","1KG_trio",cohort)) %>%
  mutate(hap1=str_replace(hap1,"-","_")) %>%
  mutate(hap2=str_replace(hap2,"-","_")) %>%
  filter(cohort %in% c("Allentoft", "1KG", "1KG_trio", "Excoffier", "SGDP", "HGDP","GTEx")) 

t_samples = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/read_depth_genotyping/genotyping/output/genotypes.likelihoods.tsv",header=TRUE) %>% 
  dplyr::select(sample) %>% 
  separate(sample,c("cohort","sample","p1","p2"),sep = "\\.") %>% 
  unique() 


t_cp = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/read_depth_genotyping/genotyping/output/genotypes.likelihoods.tsv",header=TRUE) %>% 
  #separate(sample,c("cohort","sample","p1","p2"),sep = "\\.") %>% 
  group_by(sample,label) %>% 
      filter(r==max(r)) %>%
  dplyr::select(sample,cp,label) %>%
  separate(sample,c("cohort","sample","p1","p2"),sep = "\\.") %>%
  mutate(cohort = ifelse(p2=="HunterGathererE","Hunter-GathererE",cohort))


t_grouped = t %>%
  inner_join(t_clusts %>% transmute(hap1=chrom, hap1_clust = haplotype),on="hap1") %>%
  inner_join(t_clusts %>% transmute(hap2=chrom, hap2_clust = haplotype),on="hap2") %>%
  dplyr::select(-c(cohort)) %>%
  inner_join(t_samples,on="sample") %>%
  dplyr::select(sample,cohort,hap1_clust,hap2_clust,p1,p2,best_score) %>%
  pivot_longer(cols=c("hap1_clust","hap2_clust"),names_to = "hap", values_to = "cluster") 
  

g=ggplot(t)
g+geom_histogram(aes(x=best_score))+
  theme_bw()+
  facet_wrap(~cohort,scales="free_y")

t %>% 
  group_by(cohort) %>%
  summarize(n=n())

t_cp %>% filter(label=="AMY1") %>%
  group_by(cohort) %>%
  summarize(n=n())

unique(t$cohort)
t %>% filter(cohort=="Allentoft")

unique(t_grouped$cohort)

dim(t_grouped %>% filter(cohort == "StoneAgeAncient"))/2



```

```{r}
#install.packages("ggbreak")
library(ggbreak)

#plot haplotype frequencies worldwide
#unique(t$hap1)

#### PLOT SORTED BY HAPLOTYPE copy number




clust_to_copy = t_clusts %>% 
  dplyr::select(AMY1,AMY2A,AMY2Ap,AMY2B,ID,haplotype) %>% 
  mutate(cluster=haplotype) %>%
  unique()

# #https://stackoverflow.com/questions/52214071/how-to-order-data-by-value-within-ggplot-facets

haplotype_pred_copy = t_grouped %>% 
  filter(cohort %in% c("SGDP","1KG","1KG_trio","HGDP")) %>%
  inner_join(clust_to_copy,by="cluster") %>%
  group_by(sample,cohort,p1,p2) %>%
  summarize(pred_AMY1 = sum(AMY1),
            pred_AMY2A = sum(AMY2A),
            pred_AMY2B = sum(AMY2B))

g=ggplot(t_grouped %>% filter(cohort %in% c("StoneAgeAncient","ExcoffierAncient")))
g+geom_histogram(aes(x=best_score))


BEST_SCORE_CUTOFF = 0.7
haplotype_pred_copy_ancient = t_grouped %>% 
  filter(cohort %in% c("StoneAgeAncient","ExcoffierAncient")) %>%
  filter(best_score>=BEST_SCORE_CUTOFF) %>%
  inner_join(clust_to_copy,by="cluster") %>%
  group_by(sample,cohort,p1,p2) %>%
  summarize(pred_AMY1 = sum(AMY1),
            pred_AMY2A = sum(AMY2A),
            pred_AMY2B = sum(AMY2B))

#t_sum %>% filter(p2)

t_grouped %>% filter(sample=="HG01361")

unique(t_grouped$cohort)

t_grouped %>% 
  filter(cohort %in% c("StoneAgeAncient","ExcoffierAncient")) 
  

```



```{r}
library(ggforce)
#install.packages("ggforce")

#haplotype_vs_readdepth = inner_join(haplotype_pred_copy, t_cp %>% filter(label=="AMY1") %>% dplyr::select(cp,sample), on="sample")
haplotype_vs_readdepth = inner_join(haplotype_pred_copy, t_cp %>% mutate(label=paste("cp_",label,sep="")) %>% pivot_wider(names_from=label,values_from=cp) , on="sample") 
                          #filter(cohort %in% c("1KG","1KG_trio"))
#https://stackoverflow.com/questions/61367186/pivot-longer-into-multiple-columns
longer = pivot_longer(haplotype_vs_readdepth, -c("sample","cohort","p1","p2"), names_pattern = "(.*)_(...*)$", names_to = c("type", "name")) 
tidy_hap_vs_rd = pivot_wider(longer, id_cols=c(sample,cohort,p1,p2,name),names_from=type,values_from=value)

haplotype_vs_readdepth_ancient = inner_join(haplotype_pred_copy_ancient, t_cp %>% mutate(label=paste("cp_",label,sep="")) %>% pivot_wider(names_from=label,values_from=cp) , on="sample") 
longer_anc = pivot_longer(haplotype_vs_readdepth_ancient, -c("sample","cohort","p1","p2"), names_pattern = "(.*)_(...*)$", names_to = c("type", "name")) 
tidy_hap_vs_rd_anc = pivot_wider(longer_anc, id_cols=c(sample,cohort,p1,p2,name),names_from=type,values_from=value)


stats = tidy_hap_vs_rd %>%
  mutate(corr = pred==cp) %>%
  group_by(name) %>%
  summarize(corr = sum(corr),
            n=n()) %>%
  mutate(prop=corr/n) %>%
  mutate(label=paste(signif(prop*100,2),"% (",corr,"/",n,")",sep="")) %>%
  mutate(x=ifelse(name=="AMY2A",0,2),
         y=ifelse(name=="AMY1",20,7))

stats_anc = tidy_hap_vs_rd_anc %>%
  mutate(corr = pred==cp) %>%
  group_by(name) %>%
  summarize(corr = sum(corr),
            n=n()) %>%
  mutate(prop=corr/n) %>%
  mutate(label=paste(signif(prop*100,2),"% (",corr,"/",n,")",sep="")) %>%
  mutate(x=ifelse(name=="AMY2A",0,2),
         y=ifelse(name=="AMY1",20,7))


plot_sum = tidy_hap_vs_rd %>%
            group_by(name) %>%
            mutate(n_total=n()) %>%
            group_by(name,pred,cp,n_total) %>%
            summarize(n = n()) %>%
            mutate(prop=n/n_total)
       
dummy_data = data.frame(name=c("AMY1","AMY2A","AMY2A","AMY2B"),
                        pred=c(20,0,7,7),
                        cp=c(20,0,7,7))            

g=ggplot(plot_sum)
g=g+
  geom_abline(slope=1,color="lightblue",size=.5)+
  geom_point(aes(x=cp,
                   y=pred),
             position=position_jitternormal(),
             alpha=0.2,
             size=0.05,
             data=tidy_hap_vs_rd)+
  theme_bw(base_size=7)+
  facet_wrap(~name,scales="free")+
  scale_size_continuous(range=c(0.01,2))+
  scale_x_continuous("copy number (read depth)", breaks=seq(0,20,2))+
  scale_y_continuous("copy number\n(haplotype deconvolution)", breaks=seq(0,20,2))+
  theme(strip.background=element_blank(),
        panel.grid=element_blank())+
  #theme(legend.position = "none")+
  geom_blank(aes(x=cp,y=pred),data=dummy_data)+
  geom_text(aes(x=x,y=y,label=label),data=stats,size=2,hjust=0)

fout = paste(outdir,"/copy_number_comparison.pdf",sep="")
pdf(fout, width=4 ,height=1.55)
print(g)
dev.off()

g=ggplot(tidy_hap_vs_rd_anc)
g=g+geom_abline(slope=1,color="lightblue",size=.5)+
  geom_point(aes(x=cp,
                   y=pred),
             position=position_jitternormal(),
             alpha=0.2,
             size=0.05)+
  theme_bw(base_size=7)+
  facet_wrap(~name,scales="free")+
  scale_size_continuous(range=c(0.01,2))+
  scale_x_continuous("copy number (read depth)", breaks=seq(0,20,2))+
  scale_y_continuous("copy number\n(haplotype deconvolution)", breaks=seq(0,20,2))+
  theme(strip.background=element_blank(),
        panel.grid=element_blank())+
  #theme(legend.position = "none")+
  geom_blank(aes(x=cp,y=pred),data=dummy_data)+
  geom_text(aes(x=x,y=y,label=label),data=stats_anc,size=2,hjust=0)

fout = paste(outdir,"/copy_number_comparison_ancient.pdf",sep="")
pdf(fout, width=4 ,height=1.55)
print(g)
dev.off()

print(g)

tidy_hap_vs_rd_anc %>% 
  dplyr::select(sample,p2) %>%
  unique() %>%
  group_by(p2) %>%
  summarize(n=n())

```


```{r}

colors = c("#EDB829", "#94B669", "#E97C07", "#00496F", "#0C7996","purple", "#DD4124")

outdir="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/graph_genotyping/assess_graph_genotypes/output"


t_sum_ww = t_grouped %>%
          filter(cohort %in% c("SGDP","1KG","1KG_trio","HGDP")) %>%
          ungroup() %>%
          mutate(n_total = n()) %>%
          group_by(cluster,n_total) %>%
          summarize(count = n()) %>%
          mutate(proportion=count/n_total)

write.table(t_sum_ww,
            file=paste(outdir,"/world_wide_cluster_freqs.tsv",sep=""),
            sep="\t",
            quote=FALSE,
            row.names = FALSE)

# t_sum_sub = t_grouped %>%
#           filter(cohort %in% c("SGDP","1KG","1KG_trio","HGDP")) %>%
#           group_by(p1) %>%
#           mutate(n_total = n()) %>%
#           group_by(p1, cluster,n_total) %>%
#           summarize(count = n()) %>%
#           mutate(proportion=count/n_total)

t_sum = t_grouped %>%
          filter(cohort %in% c("SGDP","1KG","HGDP")) %>%
          group_by(p2) %>%
          mutate(n_total = n()) %>%
          group_by(p2, cluster,n_total) %>%
          summarize(count = n()) %>%
          mutate(proportion=count/n_total)
t_grouped %>%
  filter(cohort %in% c("SGDP","1KG","HGDP")) %>% 
  ungroup() %>%
  summarize(n=n())

t_sum = t_sum %>% dplyr::select(p2,cluster,proportion) %>%
  pivot_wider(names_from = cluster, values_from = proportion) %>%
  pivot_longer(cols = -c(p2),names_to = "cluster",values_to="proportion") %>%
  mutate(proportion = ifelse(is.na(proportion),0,proportion)) %>%
  ungroup() %>%
  group_by(cluster) %>%
  arrange(proportion) %>%
  mutate(order=row_number())

# hap_names = data.frame(cluster=seq(0,10),
#                        hap_names = c("H7", #0,
#                                      "H3^r", #1,
#                                      "H2A2B2", #2,
#                                      "H5", #3,
#                                      "H1^a", #4,
#                                      "H3A3B3", #5,
#                                      "H9", #6,
#                                      "H4A2B2", #7,
#                                      "H2A0", #8,
#                                      "H3.2", #9,
#                                      "H4A2" #10
#                        )) %>%
#   inner_join(t_sum_ww,by="cluster") %>%
#   mutate(ww_prop = proportion) %>%
#   dplyr::select(-proportion)
  
#t_sum$cluster = factor(t_sum$cluster, levels=c(1,3,4,0,10,7,6,8,2,5,9))
#H7, H3, H2, H5, H1, H3, H9, H4, H2, H3, H4

hacked_min_max = data.frame(cluster = c("H3^r","H5","H1^a","H7","H4A2","H4A2B2","H9","H2A0","H2A2B2","H3A3B3","H1.2"),
                            mxs = c(0.55,0.55,0.55,0.1,0.1,0.1,0.1,0.1,0.1,0.1,.1),
                            line = rep(0.1,11))

t_plot = t_sum %>% mutate(hap_names=cluster)
#t_plot = inner_join(t_sum %>% mutate(cluster=as.numeric(cluster)), hap_names, by="cluster")

#t_plot$cluster = factor(t_plot$cluster, levels=c(1,3,4,0,10,7,6,8,2,5,9))
#t_plot$hap_names = factor(t_plot$hap_names, levels=(hap_names %>% arrange(-ww_prop))$hap_names)
#hacked_min_max$hap_names = factor(hacked_min_max$hap_names, levels=(hap_names %>% arrange(-ww_prop))$hap_names)


t_plot %>% arrange(hap_names)
#filter ultra rare
#t_plot = t_plot %>% filter(!hap_names %in% c("H1.2"))
t_plot = t_plot %>% filter(!hap_names %in% c("H3A2","H1.2"))

#H3A2 is found in 4 haplotypes 4/7188
4/7188*100
###


plot_order = (t_plot %>% ungroup() %>% group_by(cluster) %>%
    summarize(mu = mean(proportion)) %>%
    arrange(-mu))$cluster
  
t_plot$hap_names = factor(t_plot$hap_names,levels=plot_order)


g=ggplot(t_plot)
g=g+geom_bar(aes(y=order,fill=p2,x=proportion),stat='identity',position='dodge',alpha=.8)+
  geom_text(aes(y=order,label=p2,x=proportion),size=.75,hjust=0,position=position_nudge(x=0.001,y=0))+
  theme_bw(base_size=4)+
  scale_fill_manual(values=colors)+
  facet_wrap(~hap_names,scales="free",ncol=1,strip.position = "right",labeller = label_parsed)+
  geom_vline(xintercept=0.1,linetype="dotted",size=.25)+
  #geom_text(aes(x=mxs-.1,y=0,label=hap_names),size=1,hjust=0,data=hacked_min_max,parse=TRUE)+
  #facet_wrap(~cluster,nrow=2)+
  theme(strip.background = element_blank(),
        strip.text = element_text(angle=0))+
        #axis.text.y = element_text(angle=90,hjust=0))+
  theme(legend.position="None")+
  geom_blank(aes(x=mxs),data=hacked_min_max)+
  scale_y_discrete("")+
  scale_x_continuous("allele frequency")


fout = paste(outdir,"/hap_proportions_v3.pdf",sep="")
pdf(fout, width=1.25 ,height=3)
print(g)
dev.off()

fout = paste(outdir,"/hap_proportions_v4.pdf",sep="")
pdf(fout, width=2 ,height=2.5)
print(g+facet_wrap(~hap_names,scales="free",ncol=2))
dev.off()
print(g)

#g=ggplot(t_plot %>% filter(hap_names %in% c("H3^r","H5","H1^a")))
g=ggplot(t_plot)
g=g+geom_bar(aes(x=order,fill=p2,y=proportion),stat='identity',position='dodge',alpha=.8)+
  geom_text(aes(x=order,label=p2,y=proportion),size=.75,hjust=0.5,position=position_nudge(y=0.01,x=0),angle=0)+
  #geom_text(aes(x=0,y=mxs-0.01,label=hap_names),size=2,hjust=0,data=hacked_min_max %>% filter(!hap_names %in% c("H3^r","H5","H1^a")),parse=TRUE)+
  geom_text(aes(x=0.5,y=0.5,label=cluster),size=2,hjust=0,data=hacked_min_max,parse=TRUE)+
  theme_bw(base_size=6)+
  scale_fill_manual(values=colors)+
  facet_wrap(~hap_names,scales="free_x",ncol=2,strip.position = "bottom",labeller = label_parsed)+
  #facet_grid(.~hap_names,scales="free_x",labeller = labeller(type = label_parsed))+
  #geom_hline(yintercept=0.1,linetype="dashed",size=.25)+
  #facet_wrap(~cluster,nrow=2)+
  theme(strip.background = element_blank(),
        strip.text=element_blank())+
  theme(legend.position="None")+
  #geom_blank(aes(y=mxs),data=hacked_min_max)+
  scale_x_discrete("")+
  scale_y_continuous("allele frequency")


fout = paste(outdir,"/hap_proportions_v5.pdf",sep="")
pdf(fout, width=1.5 ,height=4)
print(g)
dev.off()

print(g)


t_sum %>% ungroup() %>% summarize(n=sum(count))

```
```{r}

###MCCARROLL COMPARISON
###SUPER COMPARISON
t_sum = t_grouped %>%
          filter(cohort %in% c("1KG","HGDP","SGDP")) %>%
          group_by(p1,p2) %>%
          mutate(n_total = n()) %>%
          group_by(p1,p2, cluster,n_total) %>%
          summarize(count = n()) %>%
          mutate(proportion=count/n_total) %>%
  mutate(hap_names=cluster) %>%
  #inner_join(hap_names, by="cluster") %>%
  ungroup() %>%
  group_by(hap_names) %>%
  arrange(proportion) %>%
  mutate(order=n())


t_data = t_grouped %>%
          mutate(hap_names=cluster) %>%
          dplyr::select(sample,cohort, p1, p2, best_score, hap, cluster, hap_names)


write.table(t_data, file="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/graph_genotyping/assess_graph_genotypes/output/haplotype_deconvolution.tsv",sep="\t",row.names=FALSE,quote=FALSE)


usher_hap_freqs = data.frame(hap_names = c("H1^a",
                                           "H3^r",
                                           "H5",
                                           "H7",
                                           "H2A0",
                                           "H4A2",
                                           "H2A2B2",
                                           "H4A2B2"),
                             usher_freqs = c(0.16,0.4,0.23,0.1,0.04,0.01, 0.02,0.04))

usher_vs_us = t_sum %>% filter(p1=="CEU") %>%
     inner_join(usher_hap_freqs,by="hap_names")

g=ggplot(usher_vs_us)
g+geom_point(aes(x=proportion,y=usher_freqs))+
  geom_text(aes(x=proportion,y=usher_freqs,label=hap_names))+
  geom_abline(slope=1)+
  theme_bw()

fout = paste(outdir,"/McCaroll_Comparison.pdf",sep="")
pdf(fout, width=1.5 ,height=4)
print(g)
dev.off()

summary(lm(usher_freqs~proportion,data=usher_vs_us))

cor(usher_vs_us$usher_freqs,usher_vs_us$proportion)
cor(usher_vs_us$usher_freqs,usher_vs_us$proportion,method='spearman')

```



```{r}
outdir="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/graph_genotyping/assess_graph_genotypes/output"

t_subsistence = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/read_depth_genotyping/genotyping/metadata/populations_subsistence.tsv",header=T,sep="\t") %>% 
                dplyr::select(type,p1)


exist = c("Yakut","Aleut","Tlingit","Mansi")
#t_subsistence %>% filter(p1 %in% exist)
#all_CAS = unique(t_gt %>% ungroup() %>% separate(sample,c("dataset","ID",'p1','p2'),sep='\\.') %>% filter(p2=="CAS") %>% dplyr::select(p1))$p1

additional_info_df = data.frame(p1=c("Tubular",
                                    "Altaian",
                                    "Ulchi",
                                    "Chukchi",
                                    "EskimoSireniki",
                                    "EskimoChaplin",
                                    "EskimoNaukan",
                                    "Even",
                                    "Itelman",
                                    "Mansi"),
                               type=c("Hunting",
                                      "Pastoralism",
                                      "Fishing",
                                      "Hunting",
                                      "Hunting",
                                      "Hunting",
                                      "Hunting",
                                      "Hunting",
                                      "Fishing",
                                      "Fishing"))

### SEE PLOT_COPY DISTRIBUTION FOR details on IDs

t_subsistence = rbind(t_subsistence,additional_info_df)


t_data_by_sus = t_grouped %>% dplyr::select(sample,p1,hap,cluster) %>% #filtering out GTEx etc doesnt matter cause of inner join
          inner_join(t_subsistence, on="p1") %>%
          mutate(type=case_when(type=="Fishing" ~ "Fishing/Hunting",
                                type=="Hunting" ~ "Fishing/Hunting",
                                TRUE ~ type)) %>%
          mutate(simple_type = ifelse(type=="Agriculture",type,"Fishing/Hunting/Pastoralism")) %>%
          mutate(type=simple_type) %>%
          filter(type!="Two or more sources") %>%
          mutate(hap_names=cluster)
          #inner_join(hap_names, by="cluster")
          

t_sum_sus = t_data_by_sus %>%
            group_by(type) %>%
            mutate(n_total = n()) %>%
            group_by(type, cluster,hap_names,n_total) %>%
            summarize(count = n()) %>%
            mutate(proportion=count/n_total)


t_sum_sus = t_sum_sus %>% ungroup() %>% dplyr::select(type,hap_names,proportion) %>%
  pivot_wider(names_from = hap_names, values_from = proportion) %>%
  pivot_longer(cols = -c(type),names_to = "hap_names",values_to="proportion") %>%
  mutate(proportion = ifelse(is.na(proportion),0,proportion)) %>%
  ungroup() %>%
  group_by(hap_names) %>%
  arrange(-proportion) %>%
  mutate(order=row_number())

#t_sum_sus$type = factor(t_sum_sus$type, levels=c("Agriculture","Pastoralism","Fishing/Hunting"))
#t_sum_sus$cluster = factor(t_sum_sus$cluster, levels=c(1,3,4,0,10,7,6,8,2,5))

sus_mx_f = t_sum_sus %>% 
  group_by(hap_names) %>%
  summarize(mx_f = max(proportion))


hacked_min_max = data.frame(cluster = c(1,3,4,0,10,7,6,8,2),
                            mxs = c(0.7,0.6,0.6,0.6,0.6,0.1,0.1,0.1,0.1),
                            line = rep(0.1,9))

#hacked_min_max$hap_names = factor(hacked_min_max$hap_names, levels=(sus_mx_f %>% arrange(-mx_f))$hap_names)

t_sum_sus$hap_names = factor(t_sum_sus$hap_names, levels=(sus_mx_f %>% arrange(-mx_f))$hap_names)

g=ggplot(t_sum_sus)
g=g+geom_bar(aes(x=order,fill=type,y=proportion),stat='identity',position='dodge',alpha=0.8)+
  #geom_text(aes(x=order,label=type,y=ifelse(hap_names=="H3^r",0,proportion)),size=.85,hjust=0,angle=90,position=position_nudge(y=0.001,x=0))+
  theme_bw(base_size=5)+
  scale_fill_manual("",values=brewer.pal(7,"Spectral")[c(1,2,7)])+
  facet_wrap(~hap_names,scales="free_x",nrow=1,strip.position = "bottom",labeller = label_parsed)+
  theme(strip.background = element_blank())+
  #theme(legend.position="None")+
  theme(legend.position="bottom",
        legend.key.size=unit(.2,"cm"),
        legend.box.margin=margin(-10,-5,-5,-5))+
  scale_x_discrete("")+
  scale_y_continuous("allele frequency")

fout = paste(outdir,"/hap_proportions_by_sustinance_v2.pdf",sep="")
pdf(fout, width=3 ,height=1)
print(g)
dev.off()


```

```{r}

outdir="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/graph_genotyping/assess_graph_genotypes/output"

######################
######################
#####FILTER FOR SV "GOOD SAMPLES"
######################



lmp <- function (modelobject) {
    if (class(modelobject) != "lm") stop("Not an object of class 'lm' ")
    f <- summary(modelobject)$fstatistic
    p <- pf(f[1],f[2],f[3],lower.tail=F)
    attributes(p) <- NULL
    return(p)
}

ancient_info = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/haplotype_analysis/metadata/ancient_info.tsv",header=T,sep="\t") %>%
  mutate(ID = sample, age=ageAverage) %>%
  mutate(sample=ID) %>%
  dplyr::select(sample, age)

ancient_info_excof = read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/read_depth_genotyping/d4_cvg_analysis/bam_to_d4/meta_data/ExcoffierAncients.tsv",header=T,sep="\t") %>%
  mutate(ID = SAMPLE_ID, age=AGE_AVG_BP) %>%
  mutate(sample=ID) %>%
  dplyr::select(sample, age)

ancient_info = rbind(ancient_info, ancient_info_excof)

t_sum = t_grouped %>%
          mutate(hap_names=cluster) %>%
          filter(best_score>BEST_SCORE_CUTOFF) %>%
          filter(cohort %in% c("StoneAgeAncient","ExcoffierAncient","1KG","SGDP","HGDP")) %>%
          filter(p2 %in% c("WEA","EarlyFarmer","Hunter-GathererW","Hunter-GathererE","Yamnaya","Europe")) %>%
          mutate(p2 = ifelse(p2=="WEA","modern-WEA",p2)) %>%
          filter((cohort %in% c("1KG","SGDP","HGDP")) | (sample %in% haplotype_pred_copy_ancient$sample)) %>%
          group_by(p2) %>%
          mutate(n_total = n()) %>%
          group_by(p2, cluster,hap_names,n_total) %>%
          summarize(count = n()) %>%
          mutate(proportion=count/n_total) %>%
          #inner_join(hap_names %>% dplyr::select(cluster,hap_names),by="cluster") %>%
          ungroup() %>%
          group_by(hap_names) %>%
          arrange(-proportion) %>%
          mutate(order=row_number())

hap_order = (t_sum %>% group_by(hap_names) %>% summarize(max_freq = max(proportion)) %>% ungroup() %>% arrange(max_freq))$hap_names
library(PNWColors)

t_sum$p2 = factor(t_sum$p2, levels=c("modern-WEA", "Hunter-GathererE", "Hunter-GathererW", "EarlyFarmer","Yamnaya"))
colors = c("#94B669", pnw_palette(name="Starfish",n=4,type="discrete"))

g=ggplot(t_sum)
g=g+geom_bar(aes(x=order,fill=p2,y=proportion),stat='identity',position='dodge',alpha=0.7)+
  geom_text(aes(x=order,label=p2,y=ifelse(hap_names=="H3^r",0,proportion)),size=1,hjust=0,angle=90,position=position_nudge(y=0.001,x=0))+
  theme_bw(base_size=4)+
  scale_fill_manual(values=colors)+
  #scale_fill_manual(values=brewer.pal(7,"Spectral")[c(1,2,7)])+
  facet_wrap(~hap_names,scales="free_x",nrow=1,strip.position = "bottom",labeller = label_parsed)+
  theme(strip.background = element_blank())+
  theme(legend.position="None")+
  scale_x_discrete("")+
  scale_y_continuous("allele frequency")

fout = paste(outdir,"/ancient_hap_proportions.pdf",sep="")
pdf(fout, width=3 ,height=.75)
print(g)
dev.off()



subset = c("H5", "H3.2", "H4A2B2", "H1^a", "H2A0", "H3^r")
subset = c("H5", "H4A2B2", "H1^a", "H3^r")
g=ggplot(t_sum %>% filter(hap_names %in% subset))
g=g+geom_bar(aes(x=order,fill=p2,y=proportion),stat='identity',position='dodge',alpha=0.7)+
  geom_text(aes(x=order,label=p2,y=ifelse(hap_names=="H3^r",0,proportion)),size=1,hjust=0,angle=90,position=position_nudge(y=0.001,x=0))+
  theme_bw(base_size=4)+
  scale_fill_manual(values=colors)+
  #scale_fill_manual(values=brewer.pal(7,"Spectral")[c(1,2,7)])+
  facet_wrap(~hap_names,scales="free_x",nrow=1,strip.position = "bottom",labeller = label_parsed)+
  theme(strip.background = element_blank())+
  theme(legend.position="None")+
  scale_x_discrete("")+
  scale_y_continuous("allele frequency")

fout = paste(outdir,"/ancient_hap_proportions_subset.pdf",sep="")
pdf(fout, width=1.2 ,height=.9)
print(g)
dev.off()


t_sum %>% arrange(proportion)
```

```{r}

generation_time = 30

t_hap_by_age = t_grouped %>%
  mutate(hap_names=cluster) %>%
       filter(best_score>BEST_SCORE_CUTOFF) %>%
        filter(cohort %in% c("StoneAgeAncient","ExcoffierAncient","HGDP","SGDP")) %>%
   filter(p2 %in% c("WEA","EarlyFarmer","Hunter-GathererW","Hunter-GathererE","Yamnaya","Europe")) %>%
        left_join(ancient_info,by="sample") %>%
        mutate(age=ifelse(!cohort %in% c("StoneAgeAncient","ExcoffierAncient"),0,as.numeric(age))) %>% 
        filter(age<20000) %>%
        filter((cohort %in% c("1KG","SGDP","HGDP")) | (sample %in% haplotype_pred_copy_ancient$sample)) %>%
        #inner_join(hap_names %>% dplyr::select(cluster,hap_names),by="cluster") %>%
        mutate(age=-1*age) %>%
        mutate(time_bin = cut(age,breaks=c(-11000,-8000,-4000,-500,0),ordered_result = TRUE))
        #mutate(time_bin = ntile(age,n=3))

t_sum_hap_by_age = t_hap_by_age %>%
          group_by(time_bin) %>%
          mutate(gen_age = age/generation_time) %>%
          mutate(n_total = n(),
                 mu_gen = mean(gen_age)) %>%
          group_by(time_bin, hap_names,n_total,mu_gen) %>%
          summarize(count = n()) %>%
          mutate(proportion=count/n_total) %>%
          group_by(hap_names) %>%
          mutate(mu_freq = mean(proportion)) %>%
  mutate(adjusted_gen = round(mu_gen-min(mu_gen))+1)

g=ggplot(t_sum_hap_by_age %>% filter(mu_freq >0.01))
g=g+geom_point(aes(x=time_bin,y=proportion),size=.1)+
  geom_line(aes(x=time_bin,y=proportion,group=hap_names))+
  facet_wrap(~hap_names,nrow=1,labeller = label_parsed,strip.position = "bottom") +
  scale_x_discrete("",labels=c("<8000","8000-4000","4000-500","modern"))+
  theme_bw(base_size=4)+
  theme(strip.background=element_blank(),
        axis.text.x=element_text(angle=45,hjust=1))

fout = paste(outdir,"/ancient_hap_freqs_binned.pdf",sep="")
pdf(fout, width=3 ,height=1)
print(g)
dev.off()

t_sum_hap_by_age %>% filter(hap_names=="H5")


t_hap_by_age = t_grouped %>%
       filter(best_score>BEST_SCORE_CUTOFF) %>%
        mutate(hap_names=cluster) %>%
        filter(cohort %in% c("StoneAgeAncient","ExcoffierAncient","HGDP","SGDP")) %>%
   filter(p2 %in% c("WEA","EarlyFarmer","Hunter-GathererW","Hunter-GathererE","Yamnaya","Europe")) %>%
        left_join(ancient_info,by="sample") %>%
        mutate(age=ifelse(!cohort %in% c("StoneAgeAncient","ExcoffierAncient"),0,as.numeric(age))) %>% 
        filter(age<20000) %>%
        #inner_join(hap_names %>% dplyr::select(cluster,hap_names),by="cluster") %>%
        mutate(age=-1*age) %>%
        mutate(time_bin = cut(age,breaks=seq(-11000,0,2500),ordered_result = TRUE)) 
        #mutate(time_bin = cut(age,breaks=seq(-11000,0,1000),ordered_result = TRUE)) 
        #mutate(time_bin = cut(age,breaks=c(-11000, -10000,  -9000,  -8000,  -7000,  -6000,  -5000,  -4000,  -3000,   0),ordered_result = TRUE))
        #mutate(time_bin = cut(age,breaks=c(-11000,-8000,-4000,-500,0),ordered_result = TRUE))


t_sum_hap_by_age = t_hap_by_age %>%
          group_by(time_bin) %>%
          mutate(gen_age = age/generation_time) %>%
          mutate(n_total = n(),
                 mu_gen = mean(gen_age)) %>%
          group_by(time_bin, hap_names,n_total,mu_gen) %>%
          summarize(count = n()) %>%
          mutate(proportion=count/n_total) %>%
          group_by(hap_names) %>%
          mutate(mu_freq = mean(proportion)) %>%
          ungroup() %>%
  mutate(adjusted_gen = round(mu_gen-min(mu_gen))+1)

t_hap_by_gen = t_sum_hap_by_age %>% 
  filter(hap_names%in%c("H1^a", "H3^r", "H5","H2A2B2","H4A2B2","H7","H2A0")) %>% 
  #filter(hap_names%in%c("H5","H4A2B2")) %>% 
  #filter(hap_names%in%c("H5")) %>% 
  #filter(hap_names%in%c("H4A2B2")) %>% 
  transmute(time=adjusted_gen,count=count,n_total=n_total,hap_names=hap_names) %>%
  mutate(f=count/n_total)

t_hap_by_gen_out = t_hap_by_gen %>% 
  mutate(value = paste(count,"/",n_total,sep="")) %>%
  dplyr::select(time,hap_names,value,n_total) %>%
  pivot_wider(values_from=value,names_from=hap_names) %>%
  pivot_longer(-c(time,n_total),names_to = "hap_names") %>%
  mutate(value=ifelse(is.na(value),paste("0/",n_total,sep=""),value)) %>%
  pivot_wider(values_from=value,names_from=hap_names) %>%
  ungroup() %>%
  dplyr::select(-n_total)
  #dplyr::select(time,H5,H4A2B2)

write.table(t_hap_by_gen_out,"/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/graph_genotyping/assess_graph_genotypes/approxWF/input/hap_freqs_t.txt",row.names=FALSE,sep="\t",quote=FALSE)

#### GROUP TOGETHER THE DEL and DUP haps

t_sum_hap_by_age_simple = t_hap_by_age %>%
          group_by(time_bin) %>%
          mutate(gen_age = age/generation_time) %>%
          mutate(n_total = n(),
                 mu_gen = mean(gen_age)) %>%
          mutate(hap_names = case_when(hap_names == "H1^a" ~ "HDEL",
                                       hap_names == "H2A0" ~ "HDEL",
                                       TRUE ~ "HDUP")) %>%
          group_by(time_bin, hap_names,n_total,mu_gen) %>%
          summarize(count = n()) %>%
          mutate(proportion=count/n_total) %>%
          group_by(hap_names) %>%
          mutate(mu_freq = mean(proportion)) %>%
          ungroup() %>%
  mutate(adjusted_gen = round(mu_gen-min(mu_gen))+1) %>%
  transmute(time=adjusted_gen,count=count,n_total=n_total,hap_names=hap_names,t_yrs=mu_gen*generation_time) %>%
  mutate(f=count/n_total)

t_hap_by_gen_out_simple = t_sum_hap_by_age_simple %>% 
  dplyr::select(-t_yrs) %>%
  mutate(value = paste(count,"/",n_total,sep="")) %>%
  dplyr::select(time,hap_names,value,n_total) %>%
  pivot_wider(values_from=value,names_from=hap_names) %>%
  pivot_longer(-c(time,n_total),names_to = "hap_names") %>%
  mutate(value=ifelse(is.na(value),paste("0/",n_total,sep=""),value)) %>%
  pivot_wider(values_from=value,names_from=hap_names) %>%
  ungroup() %>%
  dplyr::select(-n_total)

write.table(t_hap_by_gen_out_simple,"/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/graph_genotyping/assess_graph_genotypes/approxWF/input/hap_freqs_SIMPLE_t.txt",row.names=FALSE,sep="\t",quote=FALSE)

g=ggplot(t_sum_hap_by_age_simple)
g+geom_line(aes(x=time,y=count/n_total,color=hap_names))+
  geom_point(aes(x=time,y=count/n_total,size=n_total,color=hap_names))

t_sum_hap_by_age %>% select(hap_names) %>% unique()

```

```{r}
#install.packages("effects")
#https://stackoverflow.com/questions/65946270/multinomial-logistic-regression-in-r
#https://stats.oarc.ucla.edu/r/dae/multinomial-logistic-regression/


#t_data = t_hap_by_age %>% mutate(hap_names=ifelse(hap_names=="H3^r","0H3^r",hap_names))
t_data = t_hap_by_age %>% 
  #mutate(hap_names=ifelse(hap_names=="H3^r","0H3^r",hap_names)) %>%
  filter(hap_names %in% c("H3^r","H1^a","H2A2B2","H2A0","H4A2B2","H5","H7"))

#This puts the reference hap first so all stat comparisons are against it (it is highest freq)
m = nnet::multinom(hap_names~age,data=t_data)
#m = nnet::multinom(hap_names~age,data=t_hap_by_age %>% mutate(hap_names=ifelse(hap_names=="H5","0H5",hap_names)))
#m = nnet::multinom(hap_names~age,data=t_hap_by_age %>% mutate(hap_names=ifelse(hap_names=="H4A2B2","0H4A2B2",hap_names)))
#m = nnet::multinom(hap_names~age,data=t_hap_by_age)
summary(m)
z = summary(m)$coefficients/summary(m)$standard.errors
p = (1 - pnorm(abs(z), 0, 1)) * 2
#epiDisplay::mlogit.display(m, decimal=3, alpha=0.05)

p_vals = as.data.frame(p) %>% 
  mutate(padj=p.adjust(age)) %>%
  mutate(sig=padj<=0.05) %>%
  add_rownames(var="hap_names")

sig = c((p_vals %>% filter(sig))$hap_names,"H1^a","H2A0")
#sig = c((p_vals %>% filter(sig))$hap_names,"H3^r")

unique(t_hap_by_age$hap_names)
breaks = seq(-12000,0,100)

dmodel = data.frame(hap_names = rep(unique(t_hap_by_age$hap_names),each=length(breaks)),age=rep(breaks,11))
pp.model = cbind(dmodel %>% dplyr::select(age),predict(m,newdata=dmodel,type="probs",se=TRUE))



plot_df = pp.model %>% 
          pivot_longer(-c(age),names_to="hap_names",values_to="freq") %>%
          mutate(hap_names=ifelse(hap_names=="0H3^r","H3^r",hap_names)) %>%
          filter(hap_names %in% sig) 

# dup_traj = plot_df %>% 
#            filter(hap_names=="H1^a") %>%
#            mutate(hap_names = "H[dup]") %>%
#            mutate(freq=1-freq)
# plot_df = rbind(plot_df,dup_traj)

fit.eff = effects::Effect("age",m,xlevels=list(age=breaks))
conf_ints = data.frame(fit.eff$model.matrix, fit.eff$lower.prob, fit.eff$upper.prob) %>%
          dplyr::select(-c(X.Intercept.)) %>%
          pivot_longer(-age,names_pattern = "(..prob)\\.(.*)",names_to=c("type","hap_names"),values_to="freq") %>%
          pivot_wider(names_from="type",values_from="freq") %>%
          mutate(hap_names = case_when(hap_names=="H1.a" ~ "H1^a",
                                       hap_names=="H3.r" ~ "H3^r",
                                       hap_names=="X0H3.r" ~ "H3^r",
                                       TRUE ~ hap_names)) %>%
          filter(hap_names %in% sig)

# dup_conf = conf_ints %>% 
#            filter(hap_names=="H1^a") %>%
#            mutate(hap_names = "H[dup]") %>%
#            mutate(L.prob=1-L.prob,U.prob=1-U.prob)
# conf_ints = rbind(conf_ints,dup_conf)

#conf_ints$hap_names = factor(conf_ints$hap_names, levels=c("H3^r", "H5", "H1^a","H4A2B2"))
#plot_df$hap_names = factor(plot_df$hap_names, levels=c("H3^r", "H5", "H1^a","H4A2B2"))

unique(conf_ints$hap_names)

# cols = c("red",pnw_palette("Sunset2",6))
# linetypes = c(3,rep(1,6))
cols = pnw_palette("Sunset2",6)
cols = c(pnw_palette("Bay",7))
linetypes = rep(1,6)


plot_df$hap_names = factor(plot_df$hap_names,levels=c("H1^a","H2A0","H3^r","H5","H7","H2A2B2","H4A2B2"))
conf_ints$hap_names = factor(conf_ints$hap_names,levels=c("H1^a","H2A0","H3^r","H5","H7","H2A2B2","H4A2B2"))

g=ggplot(plot_df)
g=g+geom_ribbon(aes(x=age,ymin=L.prob,ymax=U.prob,group=hap_names,fill=hap_names),data=conf_ints,alpha=.1)+
  geom_line(aes(x=age,y=freq,group=hap_names,color=hap_names),size=1)+
  theme_bw(base_size=8)+
  #facet_wrap(~hap_names,labeller = label_parsed,strip.position = "top",nrow=1)+
  theme(strip.background = element_blank(),
        panel.grid=element_blank(),
        legend.key.height = unit(.25,"cm"),
        legend.key.width = unit(.5,"cm"))+
  scale_x_continuous("kyr BP",breaks=seq(-12000,0,4000),labels=-1*seq(-12,0,4))+
  scale_y_continuous("haplotype frequency")+
  scale_color_manual("",values=cols,labels=scales::parse_format())+
  scale_fill_manual("",values=cols,labels=scales::parse_format())+
  theme(legend.text = element_text(hjust=0))
  #scale_linetype_manual(values=linetypes,)

outdir="/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/graph_genotyping/assess_graph_genotypes/output"

fout = paste(outdir,"/ancient_hap_freqs_cont.pdf",sep="")
pdf(fout, width=2.5 ,height=1.6)
print(g)
dev.off()





```

```{r}


### s calc

w_hat=1-.5*q*p*s-q*q*s
d = (0.5*p*p*q*s+p*q*q*s)/(1-.5*q*p*s-q*q*s)

g=30
p_now = 0.3
p_old = 0.04
d=(p_now-p_old)/(12000/g)
p=p_old
q = 1-p_old

s = 2*d / (q*(d+p)*(p+2*q))
s

```

```{r}
t_s= read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/graph_genotyping/assess_graph_genotypes/approxWF/input/haps_t_10K.txt", header=T)

t_s = t_s %>% 
  filter(index>10000) %>%
  dplyr::select(-LL) %>%
  pivot_longer(-index, values_to = "s", names_to="hap_names") %>%
  #filter(hap_names %in% c("s_H5","s_H7","s_H4A2B2"))
  filter(hap_names %in% c("s_H1.a")) %>% mutate(s=-s) %>% mutate(hap_names = "H[dup]")

mus = t_s %>% 
  mutate(hap_lab = str_replace(hap_names,"s_","")) %>%
  group_by(hap_names,hap_lab) %>%
  summarize(mu = mean(s)) %>%
  mutate(label = paste("bar(s)==",signif(mu,2),sep="")) %>%
  ungroup() %>%
  #mutate(y=row_number()*12+55) %>%
  mutate(y=5) %>%
  mutate(x=0.025)

g_inset=ggplot(t_sum_hap_by_age_simple)
g_inset = g_inset+geom_line(aes(x=t_yrs/1000,y=count/n_total,color=hap_names),alpha=0.8)+
  theme_bw(base_size=4)+
  geom_point(aes(x=t_yrs/1000,y=count/n_total,size=n_total,color=hap_names),alpha=1)+
  theme(legend.position="None")+
  scale_x_continuous("kyr BP", breaks=c(-10,-5,0))+
  scale_y_continuous("")+
  theme(panel.grid=element_blank())+
  scale_color_manual(values=cols[c(1,7)])+
  scale_size_continuous(range=c(0.25,1))+
  theme(panel.background = element_rect(fill="transparent",color = NA),
        plot.background = element_rect(fill="transparent",color = NA))


data.tb <- tibble(x = 0.07, y = 100, 
                  plot = list(g_inset))

g=ggplot(t_s)
g=g+geom_density(aes(x=s,color=hap_names,fill=hap_names),alpha=0.25)+
  geom_vline(aes(xintercept=mu,color=hap_names),data=mus,size=.25,color='black',linetype="dashed")+
  geom_plot(data=data.tb, aes(x,y,label=plot),vp.width=(3/5)*1.1,vp.height=(1/2)*1.1)+
  # geom_label(aes(x=x,
  #                y=y,
  #                label=label,
  #                fill=hap_names),
  #            parse=TRUE,
  #            data=mus,
  #            size=2,
  #            label.padding = unit(0.05,"cm"),alpha=.5,hjust=1)+
  scale_x_continuous(expression(s[dup]))+
  scale_y_continuous("posterier density")+
  theme_bw(base_size=8)+
  theme(panel.grid = element_blank(),
        legend.position="None")+
  theme(panel.background = element_rect(fill='transparent'),
         plot.background = element_rect(fill='transparent', color=NA))
  

fout = paste(outdir,"/ancient_hap_selection_coeff.pdf",sep="")
pdf(fout, width=1.5 ,height=1.5)
print(g)
dev.off()

t_s %>%
  group_by(hap_names) %>%
  summarize(n=n(),
            gt0=sum(s>0)) %>%
  mutate(p = 1-(gt0/n))

1/90000
```
```{r}
#plot simple


t_s= read.table("/Users/petersudmant/Documents/science/sudmantlab/projects/ancient_DNA/amylase/amylase_diversity_project/graph_genotyping/assess_graph_genotypes/approxWF/input/haps_t_SIMPLE_10K.txt", header=T)

t_s = t_s %>% 
  filter(index>10000) %>%
  dplyr::select(-LL) %>%
  pivot_longer(-index, values_to = "s", names_to="hap_names") %>%
  filter(hap_names=="s_HDUP")


mus = t_s %>% 
  mutate(hap_lab = str_replace(hap_names,"s_","")) %>%
  group_by(hap_names,hap_lab) %>%
  summarize(mu = mean(s)) %>%
  mutate(label = paste("bar(s)==",signif(mu,2),sep="")) %>%
  ungroup() %>%
  mutate(y=row_number()*12+55) %>%
  mutate(x=0.045)

g=ggplot(t_s)
g=g+geom_density(aes(x=s,color=hap_names,fill=hap_names),alpha=0.25)+
  geom_vline(aes(xintercept=mu,color=hap_names),data=mus,size=.25,color='black',linetype="dashed")+
  geom_label(aes(x=x,
                 y=y,
                 label=label,
                 fill=hap_names),
             parse=TRUE,
             data=mus,
             size=2,
             label.padding = unit(0.05,"cm"),alpha=.5,hjust=1)+
  scale_x_continuous(expression(s[dup]))+
  scale_y_continuous("density")+
  theme_bw(base_size=8)+
  theme(panel.grid = element_blank(),
        legend.position="None")+
  theme(panel.background = element_rect(fill='transparent'),
         plot.background = element_rect(fill='transparent', color=NA))+
  coord_cartesian(xlim=c(0,0.05))
  

fout = paste(outdir,"/ancient_hap_selection_coeff_simple.pdf",sep="")
pdf(fout, width=1.5 ,height=1.5)
print(g)
dev.off()

t_s %>%
  group_by(hap_names) %>%
  summarize(n=n(),
            gt0=sum(s>0)) %>%
  mutate(p = 1-(gt0/n))




```

```{r}
# g=ggplot(t_s)
# g+geom_density(aes(x=s_H4A2B2))+
#   theme_bw()


# g=ggplot(t_s)
# g+geom_histogram(aes(x=s_H5))+
#   theme_bw()
# 
# g=ggplot(t_s)
# g+stat_summary(aes(x=1,y=s_H5))+
#   theme_bw()


# g=ggplot(t_s)
# g+geom_line(aes(x=index,y=LL),size=0.1)+
#   theme_bw()
# 
# dim(t_s)
# 
# t_s %>% 
#   mutate(n_total=n()) %>%
#   mutate(is0 = s_H5==0) %>%
#   group_by(is0,n_total) %>%
#   summarize(count=n()) %>%
#   mutate(prop = count/n_total)
# 
# g=ggplot(t_s)
# g+geom_point(aes(x=LL,y=s_H5),size=0.1)+
#   theme_bw()
# 
# mean(t_s$s_H5)
# 
# g=ggplot(t_s)
# g+geom_point(aes(x=index,y=s_H5),size=0.1)+
#   theme_bw()


```

```{r}

###IGNORE

#install.packages("devtools")
#require(devtools)
#library(devtools)
#install_github("mathii/slattice", build_vignettes=TRUE)
library(slattice)
vignette("slattice")
vignette("timevar")


set.seed(123456)
s.true <- 0.02 #Selection coefficient
f0=0.1 #initial frequency
sample.size <- 100 #number of chromosomes samples
sample.gens <- 10 #frequency of sampling (generations)
gens <- 100 #Number of generations
states <- 2 #Number of selection states
Ne<-1000 #Effective population size
initial.s<-c(-0.05,0.05) #Initial guess for selection coefficient
f<-simulate.wright.fisher.change(Ne, 100, f0, s.true, -s.true, c(50))
obs.pts <- rep(c(sample.size, rep(0,sample.gens-1)), times=gens/sample.gens)
obs.pts[gens] <- sample.size
obs <- generate.observations.from.path(f, obs.pts)

#Run inference
res.s<-s.estimate.soft.em.s2d(obs, Ne, initial.s, tol=0.001, verbose=FALSE, viterbi=TRUE)
#Obtain posterior mode of the allele frequency trajectory 
est.f <- res.s$call$params$states[apply(res.s$call$fb$f.fb, 2, which.max)]
est.s <- apply(res.s$call$fb$s.fb, 2, which.max)

res.s$s
res.s$call

cols<-c("#00BFC4", "#F8766D")
g<-gens
plot(obs$N.A/obs$N, pch=16, bty="n", ylab="Frequency", xlab="Generation", ylim=c(0.05, 0.5))
segments(1:(g-1), f[1:(g-1)], 2:g, f[2:g], col=cols[rep(c(2,1),each=50)], lwd=2)
segments(1:(g-1), est.f[1:(g-1)], 2:g, est.f[2:g], col="black", lwd=2, lty=1)
segments(1:(g-1), est.f[1:(g-1)], 2:g, est.f[2:g], col=cols[est.s], lwd=2, lty=3)
segments(1:(g-1), res.s$call$viterbi[1:(g-1)], 2:g, res.s$call$viterbi[2:g], col=cols[res.s$call$s.viterbi], lwd=2, lty=3)
legend("topleft", c("True", "Posterior mode", "Viterbi", "Observed"), lwd=c(2,2,2,NA), lty=c(1,3,3,NA), pch=c(NA, NA, NA, 16), col=c(1, 1, 1, 1), bty="n")


############

Ne <- 1000                              #N_e
g<-100                                  #Number of generations
p0 <- 0.1                               #Initial freq
s <- 0.05                               #Selection coefficient
data<-generate.observations(Ne, g, p0, s, missing.p=0.8, size.params=list(N=100,p=0.5))
#The data is a data frame with one row per generation, and two columns nam
estimate <- estimate.s(data$obs, Ne, method="Soft EM", verbose=TRUE)

ls(estimate)
ls(estimate$call)
estimate$call$emission.func
estimate$iterations
plot(estimate$posterior)
estimate$viterbi
estimate$iter.values
estimate$posterior
ls(estimate$call)
estimate$params
"#4DAF4A"
"#CC5500"
```

```{r}


#DUNNO
t_sum_sub= t_grouped %>%
          group_by(p1) %>%
          mutate(n_total = n()) %>%
          group_by(p1, cluster,n_total) %>%
          summarize(count = n()) %>%
          mutate(proportion=count/n_total)

g=ggplot(t_sum_sub)
g+geom_bar(aes(x=p1,fill=p1,y=proportion),stat='identity',position='dodge')+
  theme_bw()+
  #scale_fill_brewer(palette="Paired")+
  #facet_wrap(~cluster,scales="free_y",nrow=2)+
  facet_wrap(~cluster,nrow=2)+
  theme(strip.background = element_blank(),
        axis.text.x = element_text(angle=90,hjust=0))+
  theme(legend.position="None")


########LOOK IN HUNTER GATHERER VS NOT

```

```{r}

g=ggplot(t_sum)
g+geom_bar(aes(x=value,fill=p2,y=count),position="fill",stat='identity')+
  theme_bw()+
  scale_fill_brewer(palette="Spectral")


```